// @carbon/react/lib/components/FluidComboBox/FluidComboBox.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

const require_runtime = require("../../_virtual/_rolldown/runtime.js");
const require_usePrefix = require("../../internal/usePrefix.js");
const require_FormContext = require("../FluidForm/FormContext.js");
const require_index = require("../ComboBox/index.js");
let classnames = require("classnames");
classnames = require_runtime.__toESM(classnames);
let react = require("react");
react = require_runtime.__toESM(react);
let prop_types = require("prop-types");
prop_types = require_runtime.__toESM(prop_types);
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/components/FluidComboBox/FluidComboBox.tsx
/**
* Copyright IBM Corp. 2022, 2026
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
const FluidComboBox = react.default.forwardRef(function FluidComboBox({ className, isCondensed, ...other }, ref) {
	const prefix = require_usePrefix.usePrefix();
	const classNames = (0, classnames.default)(`${prefix}--list-box__wrapper--fluid`, className, { [`${prefix}--list-box__wrapper--fluid--condensed`]: isCondensed });
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_FormContext.FormContext.Provider, {
		value: { isFluid: true },
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_index.default, {
			ref,
			className: classNames,
			...other
		})
	});
});
FluidComboBox.propTypes = {
	className: prop_types.default.string,
	direction: prop_types.default.oneOf(["top", "bottom"]),
	disabled: prop_types.default.bool,
	id: prop_types.default.string.isRequired,
	initialSelectedItem: prop_types.default.oneOfType([
		prop_types.default.object,
		prop_types.default.string,
		prop_types.default.number
	]),
	invalid: prop_types.default.bool,
	invalidText: prop_types.default.node,
	isCondensed: prop_types.default.bool,
	itemToElement: prop_types.default.func,
	itemToString: prop_types.default.func,
	items: prop_types.default.array.isRequired,
	label: prop_types.default.node.isRequired,
	onChange: prop_types.default.func.isRequired,
	selectedItem: prop_types.default.oneOfType([
		prop_types.default.object,
		prop_types.default.string,
		prop_types.default.number
	]),
	titleText: prop_types.default.node,
	translateWithId: prop_types.default.func,
	warn: prop_types.default.bool,
	warnText: prop_types.default.node
};
//#endregion
exports.default = FluidComboBox;
