// @carbon/react/lib/components/FluidSelect/FluidSelect.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

const require_runtime = require("../../_virtual/_rolldown/runtime.js");
const require_usePrefix = require("../../internal/usePrefix.js");
const require_FormContext = require("../FluidForm/FormContext.js");
const require_index = require("../Select/index.js");
let classnames = require("classnames");
classnames = require_runtime.__toESM(classnames);
let react = require("react");
react = require_runtime.__toESM(react);
let prop_types = require("prop-types");
prop_types = require_runtime.__toESM(prop_types);
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/components/FluidSelect/FluidSelect.tsx
/**
* Copyright IBM Corp. 2022, 2026
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
const FluidSelect = react.default.forwardRef(({ className, children, ...other }, ref) => {
	const classNames = (0, classnames.default)(`${require_usePrefix.usePrefix()}--select--fluid`, className);
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_FormContext.FormContext.Provider, {
		value: { isFluid: true },
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_index.default, {
			ref,
			className: classNames,
			...other,
			children
		})
	});
});
FluidSelect.propTypes = {
	children: prop_types.default.node,
	className: prop_types.default.string,
	defaultValue: prop_types.default.any,
	disabled: prop_types.default.bool,
	id: prop_types.default.string.isRequired,
	invalid: prop_types.default.bool,
	invalidText: prop_types.default.node,
	labelText: prop_types.default.node,
	onChange: prop_types.default.func,
	warn: prop_types.default.bool,
	warnText: prop_types.default.node,
	readOnly: prop_types.default.bool
};
//#endregion
exports.default = FluidSelect;
