// @carbon/react/lib/components/FluidTextInput/FluidPasswordInput.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

const require_runtime = require("../../_virtual/_rolldown/runtime.js");
const require_usePrefix = require("../../internal/usePrefix.js");
const require_FormContext = require("../FluidForm/FormContext.js");
const require_PasswordInput = require("../TextInput/PasswordInput.js");
let classnames = require("classnames");
classnames = require_runtime.__toESM(classnames);
let react = require("react");
react = require_runtime.__toESM(react);
let prop_types = require("prop-types");
prop_types = require_runtime.__toESM(prop_types);
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/components/FluidTextInput/FluidPasswordInput.tsx
/**
* Copyright IBM Corp. 2016, 2025
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
const FluidPasswordInput = ({ className, ...other }) => {
	const classNames = (0, classnames.default)(className, `${require_usePrefix.usePrefix()}--text-input--fluid`);
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_FormContext.FormContext.Provider, {
		value: { isFluid: true },
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_PasswordInput.default, {
			className: classNames,
			...other
		})
	});
};
FluidPasswordInput.propTypes = {
	className: prop_types.default.string,
	defaultValue: prop_types.default.oneOfType([prop_types.default.string, prop_types.default.number]),
	disabled: prop_types.default.bool,
	hidePasswordLabel: prop_types.default.string,
	id: prop_types.default.string.isRequired,
	invalid: prop_types.default.bool,
	invalidText: prop_types.default.node,
	isPassword: prop_types.default.bool,
	labelText: prop_types.default.node.isRequired,
	onChange: prop_types.default.func,
	onClick: prop_types.default.func,
	onTogglePasswordVisibility: prop_types.default.func,
	placeholder: prop_types.default.string,
	showPasswordLabel: prop_types.default.string,
	value: prop_types.default.oneOfType([prop_types.default.string, prop_types.default.number]),
	warn: prop_types.default.bool,
	warnText: prop_types.default.node,
	readOnly: prop_types.default.bool
};
//#endregion
exports.default = FluidPasswordInput;
