// @carbon/utilities/lib/index.js
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const require_dateTimeFormat = require("./chunk-D5n0j0X6.js");
const require_carousel = require("./chunk-DVu1wdA8.js");
const require_datePartsOrder_index = require("./datePartsOrder/index.js");
const require_documentLang_index = require("./documentLang/index.js");
const require_makeDraggable_index = require("./makeDraggable/index.js");
const require_overflowHandler_index = require("./overflowHandler/index.js");
exports.createOverflowHandler = require_overflowHandler_index.createOverflowHandler;
exports.datePartsOrder = require_datePartsOrder_index.datePartsOrder;
exports.dateTimeFormat = require_dateTimeFormat.dateTimeFormat;
exports.getDocumentLang = require_documentLang_index.getDocumentLang;
exports.getSize = require_overflowHandler_index.getSize;
exports.initCarousel = require_carousel.initCarousel;
exports.makeDraggable = require_makeDraggable_index.makeDraggable;
exports.subscribeDocumentLangChange = require_documentLang_index.subscribeDocumentLangChange;
exports.updateOverflowHandler = require_overflowHandler_index.updateOverflowHandler;
var _internationalized_number = require("@internationalized/number");
Object.keys(_internationalized_number).forEach(function(k) {
	if (k !== "default" && !Object.prototype.hasOwnProperty.call(exports, k)) Object.defineProperty(exports, k, {
		enumerable: true,
		get: function() {
			return _internationalized_number[k];
		}
	});
});
