// @carbon/react/lib/components/Tooltip/DefinitionTooltip.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

const require_runtime = require("../../_virtual/_rolldown/runtime.js");
const require_usePrefix = require("../../internal/usePrefix.js");
const require_keys = require("../../internal/keyboard/keys.js");
const require_match = require("../../internal/keyboard/match.js");
const require_useId = require("../../internal/useId.js");
const require_deprecate = require("../../prop-types/deprecate.js");
const require_index = require("../Popover/index.js");
let classnames = require("classnames");
classnames = require_runtime.__toESM(classnames);
let react = require("react");
react = require_runtime.__toESM(react);
let prop_types = require("prop-types");
prop_types = require_runtime.__toESM(prop_types);
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/components/Tooltip/DefinitionTooltip.tsx
/**
* Copyright IBM Corp. 2016, 2026
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
const DefinitionTooltip = ({ align = "bottom", autoAlign, className, children, definition, defaultOpen = false, id, openOnHover, tooltipText, triggerClassName, ...rest }) => {
	const [isOpen, setOpen] = (0, react.useState)(defaultOpen);
	const prefix = require_usePrefix.usePrefix();
	const tooltipId = require_useId.useFallbackId(id);
	function onKeyDown(event) {
		if (isOpen && require_match.match(event, require_keys.Escape)) {
			event.stopPropagation();
			setOpen(false);
		}
	}
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)(require_index.Popover, {
		align,
		className,
		autoAlign,
		dropShadow: false,
		highContrast: true,
		onMouseLeave: () => {
			setOpen(false);
		},
		onMouseEnter: () => {
			if (openOnHover) setOpen(true);
		},
		onFocus: () => {
			setOpen(true);
		},
		open: isOpen,
		children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("button", {
			...rest,
			className: (0, classnames.default)(`${prefix}--definition-term`, triggerClassName),
			"aria-controls": tooltipId,
			"aria-describedby": tooltipId,
			"aria-expanded": isOpen,
			onBlur: () => {
				setOpen(false);
			},
			onMouseDown: (event) => {
				if (event.button === 0) setOpen(!isOpen);
			},
			onKeyDown,
			type: "button",
			children
		}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_index.PopoverContent, {
			className: `${prefix}--definition-tooltip`,
			id: tooltipId,
			children: tooltipText ?? definition
		})]
	});
};
DefinitionTooltip.propTypes = {
	align: prop_types.default.oneOf([
		"top",
		"top-left",
		"top-right",
		"bottom",
		"bottom-left",
		"bottom-right",
		"left",
		"left-bottom",
		"left-top",
		"right",
		"right-bottom",
		"right-top",
		"top-start",
		"top-end",
		"bottom-start",
		"bottom-end",
		"left-end",
		"left-start",
		"right-end",
		"right-start"
	]),
	autoAlign: prop_types.default.bool,
	children: prop_types.default.node.isRequired,
	className: prop_types.default.string,
	defaultOpen: prop_types.default.bool,
	definition: prop_types.default.node.isRequired,
	id: prop_types.default.string,
	openOnHover: prop_types.default.bool,
	tooltipText: require_deprecate.deprecate(prop_types.default.node, "The tooltipText prop has been deprecated. Please use the `definition` prop instead."),
	triggerClassName: prop_types.default.string
};
//#endregion
exports.DefinitionTooltip = DefinitionTooltip;
