// @carbon/react/lib/components/FluidNumberInput/FluidNumberInput.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

const require_runtime = require("../../_virtual/_rolldown/runtime.js");
const require_usePrefix = require("../../internal/usePrefix.js");
const require_FormContext = require("../FluidForm/FormContext.js");
const require_NumberFormatPropTypes = require("../NumberInput/NumberFormatPropTypes.js");
const require_NumberInput = require("../NumberInput/NumberInput.js");
let classnames = require("classnames");
classnames = require_runtime.__toESM(classnames);
let react = require("react");
react = require_runtime.__toESM(react);
let prop_types = require("prop-types");
prop_types = require_runtime.__toESM(prop_types);
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/components/FluidNumberInput/FluidNumberInput.tsx
/**
* Copyright IBM Corp. 2022, 2025
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
const FluidNumberInput = react.default.forwardRef(({ className, ...other }, ref) => {
	const classNames = (0, classnames.default)(`${require_usePrefix.usePrefix()}--number-input--fluid`, className);
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_FormContext.FormContext.Provider, {
		value: { isFluid: true },
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_NumberInput.NumberInput, {
			ref,
			className: classNames,
			...other
		})
	});
});
FluidNumberInput.propTypes = {
	allowEmpty: prop_types.default.bool,
	className: prop_types.default.string,
	defaultValue: prop_types.default.oneOfType([prop_types.default.number, prop_types.default.string]),
	disableWheel: prop_types.default.bool,
	disabled: prop_types.default.bool,
	formatOptions: require_NumberFormatPropTypes.NumberFormatOptionsPropType,
	iconDescription: prop_types.default.string,
	id: prop_types.default.string.isRequired,
	inputMode: prop_types.default.oneOf([
		"none",
		"text",
		"tel",
		"url",
		"email",
		"numeric",
		"decimal",
		"search"
	]),
	invalid: prop_types.default.bool,
	invalidText: prop_types.default.node,
	label: prop_types.default.node,
	locale: prop_types.default.string,
	max: prop_types.default.number,
	min: prop_types.default.number,
	onChange: prop_types.default.func,
	onClick: prop_types.default.func,
	onKeyUp: prop_types.default.func,
	pattern: prop_types.default.string,
	step: prop_types.default.number,
	translateWithId: prop_types.default.func,
	type: prop_types.default.oneOf(["number", "text"]),
	value: prop_types.default.oneOfType([prop_types.default.number, prop_types.default.string]),
	warn: prop_types.default.bool,
	warnText: prop_types.default.node,
	readOnly: prop_types.default.bool
};
//#endregion
exports.default = FluidNumberInput;
