// @carbon/react/lib/components/FormGroup/FormGroup.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

const require_runtime = require("../../_virtual/_rolldown/runtime.js");
const require_usePrefix = require("../../internal/usePrefix.js");
let classnames = require("classnames");
classnames = require_runtime.__toESM(classnames);
let react = require("react");
react = require_runtime.__toESM(react);
let prop_types = require("prop-types");
prop_types = require_runtime.__toESM(prop_types);
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/components/FormGroup/FormGroup.tsx
/**
* Copyright IBM Corp. 2016, 2025
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
const FormGroup = ({ disabled = false, legendId, legendText, invalid = false, children, className, message = false, messageText = "", ...rest }) => {
	const prefix = require_usePrefix.usePrefix();
	const classNamesFieldset = (0, classnames.default)(`${prefix}--fieldset`, className);
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("fieldset", {
		disabled,
		...invalid && { "data-invalid": "" },
		className: classNamesFieldset,
		...rest,
		"aria-labelledby": rest["aria-labelledby"] || legendId,
		children: [
			/* @__PURE__ */ (0, react_jsx_runtime.jsx)("legend", {
				className: `${prefix}--label`,
				id: legendId || rest["aria-labelledby"],
				children: legendText
			}),
			children,
			message ? /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
				className: `${prefix}--form__requirements`,
				children: messageText
			}) : null
		]
	});
};
FormGroup.propTypes = {
	children: prop_types.default.node,
	className: prop_types.default.string,
	disabled: prop_types.default.bool,
	invalid: prop_types.default.bool,
	legendId: prop_types.default.node,
	legendText: prop_types.default.node.isRequired,
	message: prop_types.default.bool,
	messageText: prop_types.default.string
};
//#endregion
exports.default = FormGroup;
