// @carbon/react/lib/components/Grid/FlexGrid.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

const require_runtime = require("../../_virtual/_rolldown/runtime.js");
const require_usePrefix = require("../../internal/usePrefix.js");
const require_GridContext = require("./GridContext.js");
let classnames = require("classnames");
classnames = require_runtime.__toESM(classnames);
let react = require("react");
react = require_runtime.__toESM(react);
let prop_types = require("prop-types");
prop_types = require_runtime.__toESM(prop_types);
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/components/Grid/FlexGrid.tsx
/**
* Copyright IBM Corp. 2016, 2023
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
const FlexGrid = react.default.forwardRef(({ as, condensed = false, narrow = false, fullWidth = false, withRowGap = false, className: containerClassName, children, ...rest }, ref) => {
	const prefix = require_usePrefix.usePrefix();
	const className = (0, classnames.default)(containerClassName, {
		[`${prefix}--grid`]: true,
		[`${prefix}--grid--condensed`]: condensed,
		[`${prefix}--grid--narrow`]: narrow,
		[`${prefix}--grid--full-width`]: fullWidth,
		[`${prefix}--grid--with-row-gap`]: withRowGap
	});
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_GridContext.GridSettings, {
		mode: "flexbox",
		subgrid: false,
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(as || "div", {
			className,
			ref,
			...rest,
			children
		})
	});
});
FlexGrid.propTypes = {
	as: prop_types.default.oneOfType([prop_types.default.string, prop_types.default.elementType]),
	children: prop_types.default.node,
	className: prop_types.default.string,
	condensed: prop_types.default.bool,
	fullWidth: prop_types.default.bool,
	narrow: prop_types.default.bool,
	withRowGap: prop_types.default.bool
};
const FlexGridComponent = FlexGrid;
//#endregion
exports.FlexGridComponent = FlexGridComponent;
