// @carbon/react/lib/components/Grid/Row.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

const require_runtime = require("../../_virtual/_rolldown/runtime.js");
const require_usePrefix = require("../../internal/usePrefix.js");
let classnames = require("classnames");
classnames = require_runtime.__toESM(classnames);
let react = require("react");
react = require_runtime.__toESM(react);
let prop_types = require("prop-types");
prop_types = require_runtime.__toESM(prop_types);
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/components/Grid/Row.tsx
/**
* Copyright IBM Corp. 2016, 2026
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
function Row({ as: BaseComponent = "div", condensed = false, narrow = false, className: containerClassName, children, ...rest }) {
	const prefix = require_usePrefix.usePrefix();
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(BaseComponent, {
		className: (0, classnames.default)(containerClassName, {
			[`${prefix}--row`]: true,
			[`${prefix}--row--condensed`]: condensed,
			[`${prefix}--row--narrow`]: narrow
		}),
		...rest,
		children
	});
}
Row.propTypes = {
	as: prop_types.default.oneOfType([prop_types.default.string, prop_types.default.elementType]),
	children: prop_types.default.node,
	className: prop_types.default.string,
	condensed: prop_types.default.bool,
	narrow: prop_types.default.bool
};
//#endregion
exports.default = Row;
