// @carbon/react/lib/components/Slider/Slider.Skeleton.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

const require_runtime = require("../../_virtual/_rolldown/runtime.js");
const require_usePrefix = require("../../internal/usePrefix.js");
const require_useIsomorphicEffect = require("../../internal/useIsomorphicEffect.js");
const require_SliderHandles = require("./SliderHandles.js");
let classnames = require("classnames");
classnames = require_runtime.__toESM(classnames);
let react = require("react");
react = require_runtime.__toESM(react);
let prop_types = require("prop-types");
prop_types = require_runtime.__toESM(prop_types);
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/components/Slider/Slider.Skeleton.tsx
/**
* Copyright IBM Corp. 2016, 2023
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
const SliderSkeleton = ({ ariaLabel = "slider handle", unstable_ariaLabelHandleUpper: ariaLabelHandleUpper = "upper slider handle", hideLabel, className, twoHandles, ...rest }) => {
	const prefix = require_usePrefix.usePrefix();
	const [isRtl, setIsRtl] = (0, react.useState)(false);
	require_useIsomorphicEffect.default(() => {
		setIsRtl(document ? document.dir === "rtl" : false);
	}, []);
	const containerClasses = (0, classnames.default)(`${prefix}--slider-container`, `${prefix}--skeleton`, {
		[`${prefix}--slider-container--two-handles`]: twoHandles,
		[`${prefix}--slider-container--rtl`]: isRtl
	});
	const lowerThumbClasses = (0, classnames.default)(`${prefix}--slider__thumb`, { [`${prefix}--slider__thumb--lower`]: twoHandles });
	const upperThumbClasses = (0, classnames.default)(`${prefix}--slider__thumb`, { [`${prefix}--slider__thumb--upper`]: twoHandles });
	const lowerThumbWrapperClasses = (0, classnames.default)(`${prefix}--slider__thumb-wrapper`, { [`${prefix}--slider__thumb-wrapper--lower`]: twoHandles });
	const upperThumbWrapperClasses = (0, classnames.default)(`${prefix}--slider__thumb-wrapper`, { [`${prefix}--slider__thumb-wrapper--upper`]: twoHandles });
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
		className: (0, classnames.default)(`${prefix}--form-item`, className),
		...rest,
		children: [!hideLabel && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", { className: `${prefix}--label ${prefix}--skeleton` }), /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
			className: containerClasses,
			children: [
				/* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", { className: `${prefix}--slider__range-label` }),
				/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
					className: `${prefix}--slider`,
					children: [
						/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", { className: `${prefix}--slider__track` }),
						/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", { className: `${prefix}--slider__filled-track` }),
						/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
							className: lowerThumbWrapperClasses,
							children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
								className: lowerThumbClasses,
								children: twoHandles && !isRtl ? /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_SliderHandles.LowerHandle, { "aria-label": ariaLabel }) : twoHandles && isRtl ? /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_SliderHandles.UpperHandle, { "aria-label": ariaLabelHandleUpper }) : void 0
							})
						}),
						twoHandles ? /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
							className: upperThumbWrapperClasses,
							children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
								className: upperThumbClasses,
								children: twoHandles && !isRtl ? /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_SliderHandles.UpperHandle, { "aria-label": ariaLabelHandleUpper }) : twoHandles && isRtl ? /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_SliderHandles.LowerHandle, { "aria-label": ariaLabel }) : void 0
							})
						}) : void 0
					]
				}),
				/* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", { className: `${prefix}--slider__range-label` })
			]
		})]
	});
};
SliderSkeleton.propTypes = {
	ariaLabel: prop_types.default.string,
	unstable_ariaLabelHandleUpper: prop_types.default.string,
	className: prop_types.default.string,
	hideLabel: prop_types.default.bool,
	twoHandles: prop_types.default.bool
};
//#endregion
exports.default = SliderSkeleton;
