// @carbon/react/lib/components/FluidMultiSelect/FluidMultiSelect.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

const require_runtime = require("../../_virtual/_rolldown/runtime.js");
const require_usePrefix = require("../../internal/usePrefix.js");
const require_FormContext = require("../FluidForm/FormContext.js");
const require_FilterableMultiSelect = require("../MultiSelect/FilterableMultiSelect.js");
const require_MultiSelect = require("../MultiSelect/MultiSelect.js");
let classnames = require("classnames");
classnames = require_runtime.__toESM(classnames);
let react = require("react");
react = require_runtime.__toESM(react);
let prop_types = require("prop-types");
prop_types = require_runtime.__toESM(prop_types);
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/components/FluidMultiSelect/FluidMultiSelect.tsx
/**
* Copyright IBM Corp. 2022, 2026
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
const FluidMultiSelect = react.default.forwardRef(function FluidMultiSelect({ className, isCondensed, isFilterable, ...other }, ref) {
	const prefix = require_usePrefix.usePrefix();
	const classNames = (0, classnames.default)(`${prefix}--list-box__wrapper--fluid`, className, { [`${prefix}--list-box__wrapper--fluid--condensed`]: isCondensed });
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_FormContext.FormContext.Provider, {
		value: { isFluid: true },
		children: isFilterable ? /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_FilterableMultiSelect.FilterableMultiSelect, {
			ref,
			className: classNames,
			...other
		}) : /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_MultiSelect.MultiSelect, {
			ref,
			className: classNames,
			...other
		})
	});
});
FluidMultiSelect.propTypes = {
	className: prop_types.default.string,
	clearSelectionDescription: prop_types.default.string,
	clearSelectionText: prop_types.default.string,
	compareItems: prop_types.default.func,
	direction: prop_types.default.oneOf(["top", "bottom"]),
	disabled: prop_types.default.bool,
	downshiftProps: prop_types.default.object,
	id: prop_types.default.string.isRequired,
	initialSelectedItems: prop_types.default.array,
	invalid: prop_types.default.bool,
	invalidText: prop_types.default.node,
	isCondensed: prop_types.default.bool,
	isFilterable: prop_types.default.bool,
	itemToElement: prop_types.default.func,
	itemToString: prop_types.default.func,
	items: prop_types.default.array.isRequired,
	label: prop_types.default.node.isRequired,
	locale: prop_types.default.string,
	onChange: prop_types.default.func,
	onInputValueChange: prop_types.default.func,
	onMenuChange: prop_types.default.func,
	readOnly: prop_types.default.bool,
	selectedItems: prop_types.default.array,
	selectionFeedback: prop_types.default.oneOf([
		"top",
		"fixed",
		"top-after-reopen"
	]),
	sortItems: prop_types.default.func,
	titleText: prop_types.default.node,
	translateWithId: prop_types.default.func,
	useTitleInItem: prop_types.default.bool,
	warn: prop_types.default.bool,
	warnText: prop_types.default.node
};
//#endregion
exports.default = FluidMultiSelect;
