// @carbon/react/lib/components/OrderedList/OrderedList.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

const require_runtime = require("../../_virtual/_rolldown/runtime.js");
const require_usePrefix = require("../../internal/usePrefix.js");
let classnames = require("classnames");
classnames = require_runtime.__toESM(classnames);
let react = require("react");
react = require_runtime.__toESM(react);
let prop_types = require("prop-types");
prop_types = require_runtime.__toESM(prop_types);
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/components/OrderedList/OrderedList.tsx
/**
* Copyright IBM Corp. 2016, 2023
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
function OrderedList({ className, nested = false, native = false, isExpressive = false, ...other }) {
	const prefix = require_usePrefix.usePrefix();
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("ol", {
		className: (0, classnames.default)({
			[`${prefix}--list--ordered`]: !native,
			[`${prefix}--list--ordered--native`]: native,
			[`${prefix}--list--nested`]: nested,
			[`${prefix}--list--expressive`]: isExpressive
		}, className),
		...other
	});
}
OrderedList.propTypes = {
	children: prop_types.default.node,
	className: prop_types.default.string,
	isExpressive: prop_types.default.bool,
	native: prop_types.default.bool,
	nested: prop_types.default.bool
};
//#endregion
exports.default = OrderedList;
