// @carbon/react/es/components/TreeView/TreeView.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { usePrefix } from "../../internal/usePrefix.js";
import { ArrowDown, ArrowUp, End, Home } from "../../internal/keyboard/keys.js";
import { match, matches } from "../../internal/keyboard/match.js";
import { useId } from "../../internal/useId.js";
import { useFeatureFlag } from "../FeatureFlags/index.js";
import { useControllableState } from "../../internal/useControllableState.js";
import { DepthContext, TreeContext } from "./TreeContext.js";
import TreeNode from "./TreeNode.js";
import classNames from "classnames";
import { useEffect, useMemo, useRef, useState } from "react";
import PropTypes from "prop-types";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
//#region src/components/TreeView/TreeView.tsx
/**
* Copyright IBM Corp. 2016, 2026
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
const TreeView = ({ active: prespecifiedActive, children, className, hideLabel = false, label, multiselect = false, onActivate, onSelect, selected: preselected, size = "sm", ...rest }) => {
	const enableTreeviewControllable = useFeatureFlag("enable-treeview-controllable");
	const generatedId = useId();
	const { current: treeId } = useRef(rest.id ?? generatedId);
	const prefix = usePrefix();
	const treeClasses = classNames(className, `${prefix}--tree`, { [`${prefix}--tree--${size}`]: size !== "default" });
	const treeRootRef = useRef(null);
	const treeWalker = useRef(null);
	const controllableSelectionState = useControllableState({
		value: preselected,
		onChange: onSelect,
		defaultValue: []
	});
	const uncontrollableSelectionState = useState(preselected ?? []);
	const [selected, setSelected] = enableTreeviewControllable ? controllableSelectionState : uncontrollableSelectionState;
	const controllableActiveState = useControllableState({
		value: prespecifiedActive,
		onChange: onActivate,
		defaultValue: void 0
	});
	const uncontrollableActiveState = useState(prespecifiedActive);
	const [active, setActive] = enableTreeviewControllable ? controllableActiveState : uncontrollableActiveState;
	function resetNodeTabIndices() {
		Array.prototype.forEach.call(treeRootRef?.current?.querySelectorAll("[tabIndex=\"0\"]") ?? [], (item) => {
			item.tabIndex = -1;
		});
	}
	function handleTreeSelect(event, node) {
		const nodeId = node.id;
		if (nodeId) if (multiselect && (event.metaKey || event.ctrlKey)) {
			if (!selected.includes(nodeId)) setSelected(selected.concat(nodeId));
			else setSelected(selected.filter((selectedId) => selectedId !== nodeId));
			if (!enableTreeviewControllable) onSelect?.(event, node);
		} else {
			setSelected([nodeId]);
			setActive(nodeId);
			if (!enableTreeviewControllable) onSelect?.(event, {
				activeNodeId: nodeId,
				...node
			});
		}
	}
	useEffect(() => {
		const firstNode = treeRootRef.current?.querySelector(`.${prefix}--tree-node:not(.${prefix}--tree-node--disabled)`);
		if (firstNode instanceof HTMLElement) firstNode.tabIndex = 0;
	}, [children, prefix]);
	function handleKeyDown(event) {
		event.stopPropagation();
		if (matches(event, [
			ArrowUp,
			ArrowDown,
			Home,
			End
		])) event.preventDefault();
		if (!treeWalker.current) return;
		treeWalker.current.currentNode = event.target;
		let nextFocusNode = null;
		if (match(event, ArrowUp)) nextFocusNode = treeWalker.current.previousNode();
		if (match(event, ArrowDown)) nextFocusNode = treeWalker.current.nextNode();
		if (matches(event, [
			Home,
			End,
			{ code: "KeyA" }
		])) {
			const nodeIds = [];
			if (matches(event, [Home, End])) {
				if (multiselect && event.shiftKey && event.ctrlKey && treeWalker.current.currentNode instanceof Element && !treeWalker.current.currentNode.getAttribute("aria-disabled") && !treeWalker.current.currentNode.classList.contains(`${prefix}--tree-node--hidden`)) nodeIds.push(treeWalker.current.currentNode.id);
				while (match(event, Home) ? treeWalker.current.previousNode() : treeWalker.current.nextNode()) {
					nextFocusNode = treeWalker.current.currentNode;
					if (multiselect && event.shiftKey && event.ctrlKey && nextFocusNode instanceof Element && !nextFocusNode.getAttribute("aria-disabled") && !nextFocusNode.classList.contains(`${prefix}--tree-node--hidden`)) nodeIds.push(nextFocusNode.id);
				}
			}
			if (match(event, { code: "KeyA" }) && event.ctrlKey) {
				treeWalker.current.currentNode = treeWalker.current.root;
				while (treeWalker.current.nextNode()) if (treeWalker.current.currentNode instanceof Element && !treeWalker.current.currentNode.getAttribute("aria-disabled") && !treeWalker.current.currentNode.classList.contains(`${prefix}--tree-node--hidden`)) nodeIds.push(treeWalker.current.currentNode.id);
			}
			setSelected(selected.concat(nodeIds));
		}
		if (nextFocusNode && nextFocusNode !== event.target) {
			resetNodeTabIndices();
			if (nextFocusNode instanceof HTMLElement) {
				nextFocusNode.tabIndex = 0;
				nextFocusNode.focus();
			}
		}
		rest?.onKeyDown?.(event);
	}
	useEffect(() => {
		if (treeRootRef.current && !treeWalker.current) treeWalker.current = document.createTreeWalker(treeRootRef.current, NodeFilter.SHOW_ELEMENT, { acceptNode: function(node) {
			if (!(node instanceof Element)) return NodeFilter.FILTER_SKIP;
			if (node.classList.contains(`${prefix}--tree-node--disabled`) || node.classList.contains(`${prefix}--tree-node--hidden`)) return NodeFilter.FILTER_REJECT;
			if (node.matches(`.${prefix}--tree-node`)) return NodeFilter.FILTER_ACCEPT;
			return NodeFilter.FILTER_SKIP;
		} });
	}, [prefix]);
	const labelId = `${treeId}__label`;
	const TreeLabel = () => !hideLabel ? /* @__PURE__ */ jsx("label", {
		id: labelId,
		className: `${prefix}--label`,
		children: label
	}) : null;
	const treeContextValue = useMemo(() => ({
		active,
		multiselect,
		onActivate: setActive,
		onTreeSelect: handleTreeSelect,
		selected,
		size
	}), [
		active,
		multiselect,
		setActive,
		handleTreeSelect,
		selected,
		size
	]);
	return /* @__PURE__ */ jsxs(Fragment, { children: [/* @__PURE__ */ jsx(TreeLabel, {}), /* @__PURE__ */ jsx(TreeContext.Provider, {
		value: treeContextValue,
		children: /* @__PURE__ */ jsx(DepthContext.Provider, {
			value: 0,
			children: /* @__PURE__ */ jsx("ul", {
				...rest,
				"aria-label": hideLabel ? label : void 0,
				"aria-labelledby": !hideLabel ? labelId : void 0,
				"aria-multiselectable": multiselect || void 0,
				className: treeClasses,
				onKeyDown: handleKeyDown,
				ref: treeRootRef,
				role: "tree",
				children
			})
		})
	})] });
};
TreeView.propTypes = {
	active: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
	children: PropTypes.node,
	className: PropTypes.string,
	hideLabel: PropTypes.bool,
	label: PropTypes.string.isRequired,
	multiselect: PropTypes.bool,
	onActivate: PropTypes.func,
	onSelect: PropTypes.func,
	selected: PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.string, PropTypes.number])),
	size: PropTypes.oneOf(["xs", "sm"])
};
TreeView.TreeNode = TreeNode;
//#endregion
export { TreeView as default };
