// @carbon/react/lib/components/Breadcrumb/Breadcrumb.Skeleton.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

const require_runtime = require("../../_virtual/_rolldown/runtime.js");
const require_usePrefix = require("../../internal/usePrefix.js");
let classnames = require("classnames");
classnames = require_runtime.__toESM(classnames);
let react = require("react");
react = require_runtime.__toESM(react);
let prop_types = require("prop-types");
prop_types = require_runtime.__toESM(prop_types);
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/components/Breadcrumb/Breadcrumb.Skeleton.tsx
/**
* Copyright IBM Corp. 2016, 2025
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
function Item() {
	const prefix = require_usePrefix.usePrefix();
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: `${prefix}--breadcrumb-item`,
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", {
			className: `${prefix}--link`,
			children: "\xA0"
		})
	});
}
function BreadcrumbSkeleton({ className, items = 3, noTrailingSlash, size, ...rest }) {
	const prefix = require_usePrefix.usePrefix();
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: (0, classnames.default)({
			[`${prefix}--breadcrumb`]: true,
			[`${prefix}--skeleton`]: true,
			[`${prefix}--breadcrumb--no-trailing-slash`]: noTrailingSlash,
			[`${prefix}--breadcrumb--sm`]: size === "sm"
		}, className),
		...rest,
		children: Array.from({ length: items }, (_, i) => /* @__PURE__ */ (0, react_jsx_runtime.jsx)(Item, {}, i))
	});
}
BreadcrumbSkeleton.propTypes = {
	className: prop_types.default.string,
	items: prop_types.default.number,
	noTrailingSlash: prop_types.default.bool,
	size: prop_types.default.oneOf(["sm", "md"])
};
//#endregion
exports.default = BreadcrumbSkeleton;
