// @carbon/react/es/components/Stack/Stack.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { usePrefix } from "../../internal/usePrefix.js";
import classNames from "classnames";
import { forwardRef } from "react";
import PropTypes from "prop-types";
import { jsx } from "react/jsx-runtime";
import { spacing } from "@carbon/layout";
//#region src/components/Stack/Stack.tsx
/**
* Copyright IBM Corp. 2016, 2025
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
/**
* The steps in the spacing scale
*/
const SPACING_STEPS = Array.from({ length: spacing.length - 1 }, (_, step) => step + 1);
/**
* The Stack component is a useful layout utility in a component-based model.
* This allows components to not use margin and instead delegate the
* responsibility of positioning and layout to parent components.
*
* In the case of the Stack component, it uses the spacing scale from the
* Design Language in order to determine how much space there should be between
* items rendered by the Stack component. It also supports a custom `gap` prop
* which will allow a user to provide a custom value for the gap of the layout.
*
* This component supports both horizontal and vertical orientations.
*
* Inspiration for this component:
*
* - https://paste.twilio.design/layout/stack/
* - https://github.com/Workday/canvas-kit/blob/f2f599654876700f483a1d8c5de82a41315c76f1/modules/labs-react/layout/lib/Stack.tsx
*/
const Stack = forwardRef((props, ref) => {
	const { as: BaseComponent = "div", children, className: customClassName, gap, orientation = "vertical", ...rest } = props;
	const prefix = usePrefix();
	const className = classNames(customClassName, {
		[`${prefix}--stack-${orientation}`]: true,
		[`${prefix}--stack-scale-${gap}`]: typeof gap === "number"
	});
	const style = { ...rest.style };
	if (typeof gap === "string") style[`--${prefix}-stack-gap`] = gap;
	return /* @__PURE__ */ jsx(BaseComponent, {
		...rest,
		ref,
		className,
		style,
		children
	});
});
Stack.propTypes = {
	as: PropTypes.elementType,
	children: PropTypes.node,
	className: PropTypes.string,
	gap: PropTypes.oneOfType([PropTypes.string, PropTypes.oneOf(SPACING_STEPS)]),
	orientation: PropTypes.oneOf(["horizontal", "vertical"])
};
//#endregion
export { Stack };
