// @carbon/react/es/components/OrderedList/OrderedList.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { usePrefix } from "../../internal/usePrefix.js";
import classNames from "classnames";
import "react";
import PropTypes from "prop-types";
import { jsx } from "react/jsx-runtime";
//#region src/components/OrderedList/OrderedList.tsx
/**
* Copyright IBM Corp. 2016, 2023
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
function OrderedList({ className, nested = false, native = false, isExpressive = false, ...other }) {
	const prefix = usePrefix();
	return /* @__PURE__ */ jsx("ol", {
		className: classNames({
			[`${prefix}--list--ordered`]: !native,
			[`${prefix}--list--ordered--native`]: native,
			[`${prefix}--list--nested`]: nested,
			[`${prefix}--list--expressive`]: isExpressive
		}, className),
		...other
	});
}
OrderedList.propTypes = {
	children: PropTypes.node,
	className: PropTypes.string,
	isExpressive: PropTypes.bool,
	native: PropTypes.bool,
	nested: PropTypes.bool
};
//#endregion
export { OrderedList as default };
