// @carbon/react/es/components/TabContent/TabContent.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { usePrefix } from "../../internal/usePrefix.js";
import classNames from "classnames";
import "react";
import PropTypes from "prop-types";
import { jsx } from "react/jsx-runtime";
//#region src/components/TabContent/TabContent.tsx
/**
* Copyright IBM Corp. 2016, 2023
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
function TabContent(props) {
	const { className, selected, children, ...other } = props;
	const tabContentClasses = classNames(`${usePrefix()}--tab-content`, className);
	return /* @__PURE__ */ jsx("div", {
		role: "tabpanel",
		...other,
		className: tabContentClasses,
		selected,
		hidden: !selected,
		children
	});
}
TabContent.propTypes = {
	children: PropTypes.node,
	className: PropTypes.string,
	selected: PropTypes.bool
};
//#endregion
export { TabContent as default };
