// @carbon/react/lib/components/Theme/index.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

const require_runtime = require("../../_virtual/_rolldown/runtime.js");
const require_usePrefix = require("../../internal/usePrefix.js");
const require_LayerContext = require("../Layer/LayerContext.js");
const require_useMatchMedia = require("../../internal/useMatchMedia.js");
let classnames = require("classnames");
classnames = require_runtime.__toESM(classnames);
let react = require("react");
react = require_runtime.__toESM(react);
let prop_types = require("prop-types");
prop_types = require_runtime.__toESM(prop_types);
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/components/Theme/index.tsx
/**
* Copyright IBM Corp. 2016, 2026
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
const ThemeContext = react.default.createContext({ theme: "white" });
const GlobalTheme = react.default.forwardRef(({ children, theme }, ref) => {
	const value = (0, react.useMemo)(() => {
		return { theme };
	}, [theme]);
	const childrenWithProps = (0, react.isValidElement)(children) ? (0, react.cloneElement)(children, { ref }) : children;
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(ThemeContext.Provider, {
		value,
		children: childrenWithProps
	});
});
GlobalTheme.propTypes = {
	children: prop_types.default.node,
	theme: prop_types.default.oneOf([
		"white",
		"g10",
		"g90",
		"g100"
	])
};
/**
* Specify the theme to be applied to a page, or a region in a page
*/
function Theme({ as: BaseComponent = "div", className: customClassName, theme, ...rest }) {
	const prefix = require_usePrefix.usePrefix();
	const className = (0, classnames.default)(customClassName, {
		[`${prefix}--white`]: theme === "white",
		[`${prefix}--g10`]: theme === "g10",
		[`${prefix}--g90`]: theme === "g90",
		[`${prefix}--g100`]: theme === "g100",
		[`${prefix}--layer-one`]: true
	});
	const value = react.default.useMemo(() => {
		return {
			theme,
			isDark: theme && ["g90", "g100"].includes(theme)
		};
	}, [theme]);
	const BaseComponentAsAny = BaseComponent;
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(ThemeContext.Provider, {
		value,
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_LayerContext.LayerContext.Provider, {
			value: 1,
			children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(BaseComponentAsAny, {
				...rest,
				className
			})
		})
	});
}
Theme.propTypes = {
	as: prop_types.default.oneOfType([
		prop_types.default.func,
		prop_types.default.string,
		prop_types.default.elementType
	]),
	children: prop_types.default.node,
	className: prop_types.default.string,
	theme: prop_types.default.oneOf([
		"white",
		"g10",
		"g90",
		"g100"
	])
};
/**
* Get access to the current theme
*/
function useTheme() {
	return react.default.useContext(ThemeContext);
}
function usePrefersDarkScheme() {
	return require_useMatchMedia.useMatchMedia("(prefers-color-scheme: dark)");
}
//#endregion
exports.GlobalTheme = GlobalTheme;
exports.Theme = Theme;
exports.ThemeContext = ThemeContext;
exports.usePrefersDarkScheme = usePrefersDarkScheme;
exports.useTheme = useTheme;
