// @carbon/react/lib/components/PageHeader/index.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

const require_runtime = require("../../_virtual/_rolldown/runtime.js");
const require_PageHeader = require("./PageHeader.js");
//#region src/components/PageHeader/index.tsx
var PageHeader_exports = /* @__PURE__ */ require_runtime.__exportAll({
	BreadcrumbBar: () => require_PageHeader.BreadcrumbBar,
	Content: () => require_PageHeader.Content,
	ContentPageActions: () => require_PageHeader.ContentPageActions,
	ContentText: () => require_PageHeader.ContentText,
	HeroImage: () => require_PageHeader.HeroImage,
	PageHeader: () => require_PageHeader.PageHeader,
	PageHeaderBreadcrumbBar: () => require_PageHeader.PageHeaderBreadcrumbBar,
	PageHeaderContent: () => require_PageHeader.PageHeaderContent,
	PageHeaderContentPageActions: () => require_PageHeader.PageHeaderContentPageActions,
	PageHeaderContentText: () => require_PageHeader.PageHeaderContentText,
	PageHeaderHeroImage: () => require_PageHeader.PageHeaderHeroImage,
	PageHeaderTabBar: () => require_PageHeader.PageHeaderTabBar,
	Root: () => require_PageHeader.Root,
	TabBar: () => require_PageHeader.TabBar
});
//#endregion
Object.defineProperty(exports, "PageHeader_exports", {
	enumerable: true,
	get: function() {
		return PageHeader_exports;
	}
});
