// @carbon/react/lib/components/OverflowMenuItem/OverflowMenuItem.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

const require_runtime = require("../../_virtual/_rolldown/runtime.js");
const require_usePrefix = require("../../internal/usePrefix.js");
const require_Text = require("../Text/Text.js");
const require_keys = require("../../internal/keyboard/keys.js");
const require_match = require("../../internal/keyboard/match.js");
const require_useId = require("../../internal/useId.js");
const require_warning = require("../../internal/warning.js");
let classnames = require("classnames");
classnames = require_runtime.__toESM(classnames);
let react = require("react");
react = require_runtime.__toESM(react);
let prop_types = require("prop-types");
prop_types = require_runtime.__toESM(prop_types);
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/components/OverflowMenuItem/OverflowMenuItem.tsx
/**
* Copyright IBM Corp. 2016, 2026
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
const OverflowMenuItem = (0, react.forwardRef)((props, ref) => {
	const { className, closeMenu, dangerDescription = "", disabled = false, handleOverflowMenuItemFocus, hasDivider = false, href, isDelete = false, index, itemText = "Provide itemText", onClick = () => {}, onKeyDown = () => {}, requireTitle, title, wrapperClassName, ...rest } = props;
	const prefix = require_usePrefix.usePrefix();
	function setTabFocus(evt) {
		if (require_match.match(evt, require_keys.ArrowDown)) handleOverflowMenuItemFocus?.({
			currentIndex: index,
			direction: 1
		});
		if (require_match.match(evt, require_keys.ArrowUp)) handleOverflowMenuItemFocus?.({
			currentIndex: index,
			direction: -1
		});
	}
	function handleClick(evt) {
		onClick(evt);
		if (closeMenu) closeMenu();
	}
	require_warning.warning(!!closeMenu, "`<OverflowMenuItem>` detected missing `closeMenu` prop. `closeMenu` is required to let `<OverflowMenu>` close the menu upon actions on `<OverflowMenuItem>`. Please make sure `<OverflowMenuItem>` is a direct child of `<OverflowMenu>.");
	const overflowMenuBtnClasses = (0, classnames.default)(`${prefix}--overflow-menu-options__btn`, className);
	const overflowMenuItemClasses = (0, classnames.default)(`${prefix}--overflow-menu-options__option`, {
		[`${prefix}--overflow-menu--divider`]: hasDivider,
		[`${prefix}--overflow-menu-options__option--danger`]: isDelete,
		[`${prefix}--overflow-menu-options__option--disabled`]: disabled
	}, wrapperClassName);
	const TagToUse = href ? "a" : "button";
	const hasDangerDescription = isDelete && Boolean(dangerDescription);
	const assistiveId = require_useId.useId("danger-description");
	const OverflowMenuItemContent = (() => {
		if (typeof itemText !== "string") return itemText;
		return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)(react_jsx_runtime.Fragment, { children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
			className: `${prefix}--overflow-menu-options__option-content`,
			children: itemText
		}), hasDangerDescription && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", {
			id: assistiveId,
			className: `${prefix}--visually-hidden`,
			children: dangerDescription
		})] });
	})();
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_Text.Text, {
		as: "li",
		className: overflowMenuItemClasses,
		role: "none",
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(TagToUse, {
			className: overflowMenuBtnClasses,
			disabled,
			href,
			onClick: handleClick,
			onKeyDown: (evt) => {
				setTabFocus(evt);
				onKeyDown(evt);
			},
			role: "menuitem",
			ref,
			tabIndex: -1,
			title: requireTitle ? title || itemText : void 0,
			...rest,
			children: OverflowMenuItemContent
		})
	});
});
OverflowMenuItem.propTypes = {
	className: prop_types.default.string,
	closeMenu: prop_types.default.func,
	dangerDescription: prop_types.default.string,
	disabled: prop_types.default.bool,
	handleOverflowMenuItemFocus: prop_types.default.func,
	hasDivider: prop_types.default.bool,
	href: prop_types.default.string,
	index: prop_types.default.number,
	isDelete: prop_types.default.bool,
	itemText: prop_types.default.node.isRequired,
	onBlur: prop_types.default.func,
	onClick: prop_types.default.func,
	onFocus: prop_types.default.func,
	onKeyDown: prop_types.default.func,
	onKeyUp: prop_types.default.func,
	onMouseDown: prop_types.default.func,
	onMouseEnter: prop_types.default.func,
	onMouseLeave: prop_types.default.func,
	onMouseUp: prop_types.default.func,
	requireTitle: prop_types.default.bool,
	title: prop_types.default.string,
	wrapperClassName: prop_types.default.string
};
//#endregion
exports.default = OverflowMenuItem;
