// @carbon/react/es/components/FluidTimePickerSelect/FluidTimePickerSelect.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import FluidSelect_default from "../FluidSelect/index.js";
import React from "react";
import PropTypes from "prop-types";
import { jsx } from "react/jsx-runtime";
//#region src/components/FluidTimePickerSelect/FluidTimePickerSelect.tsx
/**
* Copyright IBM Corp. 2022, 2026
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
const FluidTimePickerSelect = React.forwardRef(({ children, className, ...other }, ref) => {
	return /* @__PURE__ */ jsx(FluidSelect_default, {
		className,
		ref,
		...other,
		children
	});
});
FluidTimePickerSelect.propTypes = {
	children: PropTypes.node,
	className: PropTypes.string,
	defaultValue: PropTypes.any,
	disabled: PropTypes.bool,
	id: PropTypes.string.isRequired,
	labelText: PropTypes.node,
	onChange: PropTypes.func,
	readOnly: PropTypes.bool
};
//#endregion
export { FluidTimePickerSelect as default };
