// @carbon/react/lib/components/ToggleSmall/ToggleSmall.Skeleton.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

const require_runtime = require("../../_virtual/_rolldown/runtime.js");
const require_usePrefix = require("../../internal/usePrefix.js");
let classnames = require("classnames");
classnames = require_runtime.__toESM(classnames);
let react = require("react");
react = require_runtime.__toESM(react);
let prop_types = require("prop-types");
prop_types = require_runtime.__toESM(prop_types);
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/components/ToggleSmall/ToggleSmall.Skeleton.tsx
/**
* Copyright IBM Corp. 2016, 2025
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
const ToggleSmallSkeleton = ({ id, labelText, className, ...rest }) => {
	const prefix = (0, react.useContext)(require_usePrefix.PrefixContext);
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
		className: (0, classnames.default)(`${prefix}--form-item`, className),
		...rest,
		children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("input", {
			type: "checkbox",
			id,
			className: `${prefix}--toggle ${prefix}--toggle--small ${prefix}--skeleton`
		}), /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("label", {
			className: `${prefix}--toggle__label ${prefix}--skeleton`,
			htmlFor: id,
			children: [labelText && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", {
				className: `${prefix}--toggle__label-text`,
				children: labelText
			}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", {
				className: `${prefix}--toggle__appearance`,
				children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("svg", {
					className: `${prefix}--toggle__check`,
					width: "6px",
					height: "5px",
					viewBox: "0 0 6 5",
					children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("path", { d: "M2.2403 2.7299L4.9245 0 6 1.1117 2.2384 5 0 2.6863 1.0612 1.511z" })
				})
			})]
		})]
	});
};
ToggleSmallSkeleton.propTypes = {
	["aria-label"]: prop_types.default.string.isRequired,
	className: prop_types.default.string,
	id: prop_types.default.string,
	labelText: prop_types.default.string
};
//#endregion
exports.ToggleSmallSkeleton = ToggleSmallSkeleton;
