// @carbon/colors/umd/index.js
(function(global, factory) {
	typeof exports === "object" && typeof module !== "undefined" ? factory(exports) : typeof define === "function" && define.amd ? define(["exports"], factory) : (global = typeof globalThis !== "undefined" ? globalThis : global || self, factory(global.CarbonColors = {}));
})(this, function(exports) {
	Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
	//#region src/colors.ts
	/**
	* Copyright IBM Corp. 2018, 2026
	*
	* This source code is licensed under the Apache-2.0 license found in the
	* LICENSE file in the root directory of this source tree.
	*/
	const black = "#000000";
	const black100 = black;
	const blackHover = "#212121";
	const white = "#ffffff";
	const white0 = white;
	const whiteHover = "#e8e8e8";
	const yellow10 = "#fcf4d6";
	const yellow20 = "#fddc69";
	const yellow30 = "#f1c21b";
	const yellow40 = "#d2a106";
	const yellow50 = "#b28600";
	const yellow60 = "#8e6a00";
	const yellow70 = "#684e00";
	const yellow80 = "#483700";
	const yellow90 = "#302400";
	const yellow100 = "#1c1500";
	const yellow = {
		10: yellow10,
		20: yellow20,
		30: yellow30,
		40: yellow40,
		50: yellow50,
		60: yellow60,
		70: yellow70,
		80: yellow80,
		90: yellow90,
		100: yellow100
	};
	const yellow10Hover = "#f8e6a0";
	const yellow20Hover = "#fccd27";
	const yellow30Hover = "#ddb00e";
	const yellow40Hover = "#bc9005";
	const yellow50Hover = "#9e7700";
	const yellow60Hover = "#755800";
	const yellow70Hover = "#806000";
	const yellow80Hover = "#5c4600";
	const yellow90Hover = "#3d2e00";
	const yellow100Hover = "#332600";
	const yellowHover = {
		10: yellow10Hover,
		20: yellow20Hover,
		30: yellow30Hover,
		40: yellow40Hover,
		50: yellow50Hover,
		60: yellow60Hover,
		70: yellow70Hover,
		80: yellow80Hover,
		90: yellow90Hover,
		100: yellow100Hover
	};
	const orange10 = "#fff2e8";
	const orange20 = "#ffd9be";
	const orange30 = "#ffb784";
	const orange40 = "#ff832b";
	const orange50 = "#eb6200";
	const orange60 = "#ba4e00";
	const orange70 = "#8a3800";
	const orange80 = "#5e2900";
	const orange90 = "#3e1a00";
	const orange100 = "#231000";
	const orange = {
		10: orange10,
		20: orange20,
		30: orange30,
		40: orange40,
		50: orange50,
		60: orange60,
		70: orange70,
		80: orange80,
		90: orange90,
		100: orange100
	};
	const orange10Hover = "#ffe2cc";
	const orange20Hover = "#ffc69e";
	const orange30Hover = "#ff9d57";
	const orange40Hover = "#fa6800";
	const orange50Hover = "#cc5500";
	const orange60Hover = "#9e4200";
	const orange70Hover = "#a84400";
	const orange80Hover = "#753300";
	const orange90Hover = "#522200";
	const orange100Hover = "#421e00";
	const orangeHover = {
		10: orange10Hover,
		20: orange20Hover,
		30: orange30Hover,
		40: orange40Hover,
		50: orange50Hover,
		60: orange60Hover,
		70: orange70Hover,
		80: orange80Hover,
		90: orange90Hover,
		100: orange100Hover
	};
	const red10 = "#fff1f1";
	const red20 = "#ffd7d9";
	const red30 = "#ffb3b8";
	const red40 = "#ff8389";
	const red50 = "#fa4d56";
	const red60 = "#da1e28";
	const red70 = "#a2191f";
	const red80 = "#750e13";
	const red90 = "#520408";
	const red100 = "#2d0709";
	const red = {
		10: red10,
		20: red20,
		30: red30,
		40: red40,
		50: red50,
		60: red60,
		70: red70,
		80: red80,
		90: red90,
		100: red100
	};
	const red100Hover = "#540d11";
	const red90Hover = "#66050a";
	const red80Hover = "#921118";
	const red70Hover = "#c21e25";
	const red60Hover = "#b81922";
	const red50Hover = "#ee0713";
	const red40Hover = "#ff6168";
	const red30Hover = "#ff99a0";
	const red20Hover = "#ffc2c5";
	const red10Hover = "#ffe0e0";
	const redHover = {
		100: red100Hover,
		90: red90Hover,
		80: red80Hover,
		70: red70Hover,
		60: red60Hover,
		50: red50Hover,
		40: red40Hover,
		30: red30Hover,
		20: red20Hover,
		10: red10Hover
	};
	const magenta10 = "#fff0f7";
	const magenta20 = "#ffd6e8";
	const magenta30 = "#ffafd2";
	const magenta40 = "#ff7eb6";
	const magenta50 = "#ee5396";
	const magenta60 = "#d02670";
	const magenta70 = "#9f1853";
	const magenta80 = "#740937";
	const magenta90 = "#510224";
	const magenta100 = "#2a0a18";
	const magenta = {
		10: magenta10,
		20: magenta20,
		30: magenta30,
		40: magenta40,
		50: magenta50,
		60: magenta60,
		70: magenta70,
		80: magenta80,
		90: magenta90,
		100: magenta100
	};
	const magenta100Hover = "#53142f";
	const magenta90Hover = "#68032e";
	const magenta80Hover = "#8e0b43";
	const magenta70Hover = "#bf1d63";
	const magenta60Hover = "#b0215f";
	const magenta50Hover = "#e3176f";
	const magenta40Hover = "#ff57a0";
	const magenta30Hover = "#ff94c3";
	const magenta20Hover = "#ffbdda";
	const magenta10Hover = "#ffe0ef";
	const magentaHover = {
		100: magenta100Hover,
		90: magenta90Hover,
		80: magenta80Hover,
		70: magenta70Hover,
		60: magenta60Hover,
		50: magenta50Hover,
		40: magenta40Hover,
		30: magenta30Hover,
		20: magenta20Hover,
		10: magenta10Hover
	};
	const purple10 = "#f6f2ff";
	const purple20 = "#e8daff";
	const purple30 = "#d4bbff";
	const purple40 = "#be95ff";
	const purple50 = "#a56eff";
	const purple60 = "#8a3ffc";
	const purple70 = "#6929c4";
	const purple80 = "#491d8b";
	const purple90 = "#31135e";
	const purple100 = "#1c0f30";
	const purple = {
		10: purple10,
		20: purple20,
		30: purple30,
		40: purple40,
		50: purple50,
		60: purple60,
		70: purple70,
		80: purple80,
		90: purple90,
		100: purple100
	};
	const purple100Hover = "#341c59";
	const purple90Hover = "#40197b";
	const purple80Hover = "#5b24ad";
	const purple70Hover = "#7c3dd6";
	const purple60Hover = "#7822fb";
	const purple50Hover = "#9352ff";
	const purple40Hover = "#ae7aff";
	const purple30Hover = "#c5a3ff";
	const purple20Hover = "#dcc7ff";
	const purple10Hover = "#ede5ff";
	const purpleHover = {
		100: purple100Hover,
		90: purple90Hover,
		80: purple80Hover,
		70: purple70Hover,
		60: purple60Hover,
		50: purple50Hover,
		40: purple40Hover,
		30: purple30Hover,
		20: purple20Hover,
		10: purple10Hover
	};
	const blue10 = "#edf5ff";
	const blue20 = "#d0e2ff";
	const blue30 = "#a6c8ff";
	const blue40 = "#78a9ff";
	const blue50 = "#4589ff";
	const blue60 = "#0f62fe";
	const blue70 = "#0043ce";
	const blue80 = "#002d9c";
	const blue90 = "#001d6c";
	const blue100 = "#001141";
	const blue = {
		10: blue10,
		20: blue20,
		30: blue30,
		40: blue40,
		50: blue50,
		60: blue60,
		70: blue70,
		80: blue80,
		90: blue90,
		100: blue100
	};
	const blue100Hover = "#001f75";
	const blue90Hover = "#00258a";
	const blue80Hover = "#0039c7";
	const blue70Hover = "#0053ff";
	const blue60Hover = "#0050e6";
	const blue50Hover = "#1f70ff";
	const blue40Hover = "#5c97ff";
	const blue30Hover = "#8ab6ff";
	const blue20Hover = "#b8d3ff";
	const blue10Hover = "#dbebff";
	const blueHover = {
		100: blue100Hover,
		90: blue90Hover,
		80: blue80Hover,
		70: blue70Hover,
		60: blue60Hover,
		50: blue50Hover,
		40: blue40Hover,
		30: blue30Hover,
		20: blue20Hover,
		10: blue10Hover
	};
	const cyan10 = "#e5f6ff";
	const cyan20 = "#bae6ff";
	const cyan30 = "#82cfff";
	const cyan40 = "#33b1ff";
	const cyan50 = "#1192e8";
	const cyan60 = "#0072c3";
	const cyan70 = "#00539a";
	const cyan80 = "#003a6d";
	const cyan90 = "#012749";
	const cyan100 = "#061727";
	const cyan = {
		10: cyan10,
		20: cyan20,
		30: cyan30,
		40: cyan40,
		50: cyan50,
		60: cyan60,
		70: cyan70,
		80: cyan80,
		90: cyan90,
		100: cyan100
	};
	const cyan10Hover = "#cceeff";
	const cyan20Hover = "#99daff";
	const cyan30Hover = "#57beff";
	const cyan40Hover = "#059fff";
	const cyan50Hover = "#0f7ec8";
	const cyan60Hover = "#005fa3";
	const cyan70Hover = "#0066bd";
	const cyan80Hover = "#00498a";
	const cyan90Hover = "#013360";
	const cyan100Hover = "#0b2947";
	const cyanHover = {
		10: cyan10Hover,
		20: cyan20Hover,
		30: cyan30Hover,
		40: cyan40Hover,
		50: cyan50Hover,
		60: cyan60Hover,
		70: cyan70Hover,
		80: cyan80Hover,
		90: cyan90Hover,
		100: cyan100Hover
	};
	const teal10 = "#d9fbfb";
	const teal20 = "#9ef0f0";
	const teal30 = "#3ddbd9";
	const teal40 = "#08bdba";
	const teal50 = "#009d9a";
	const teal60 = "#007d79";
	const teal70 = "#005d5d";
	const teal80 = "#004144";
	const teal90 = "#022b30";
	const teal100 = "#081a1c";
	const teal = {
		10: teal10,
		20: teal20,
		30: teal30,
		40: teal40,
		50: teal50,
		60: teal60,
		70: teal70,
		80: teal80,
		90: teal90,
		100: teal100
	};
	const teal10Hover = "#acf6f6";
	const teal20Hover = "#57e5e5";
	const teal30Hover = "#25cac8";
	const teal40Hover = "#07aba9";
	const teal50Hover = "#008a87";
	const teal60Hover = "#006b68";
	const teal70Hover = "#007070";
	const teal80Hover = "#005357";
	const teal90Hover = "#033940";
	const teal100Hover = "#0f3034";
	const tealHover = {
		10: teal10Hover,
		20: teal20Hover,
		30: teal30Hover,
		40: teal40Hover,
		50: teal50Hover,
		60: teal60Hover,
		70: teal70Hover,
		80: teal80Hover,
		90: teal90Hover,
		100: teal100Hover
	};
	const green10 = "#defbe6";
	const green20 = "#a7f0ba";
	const green30 = "#6fdc8c";
	const green40 = "#42be65";
	const green50 = "#24a148";
	const green60 = "#198038";
	const green70 = "#0e6027";
	const green80 = "#044317";
	const green90 = "#022d0d";
	const green100 = "#071908";
	const green = {
		10: green10,
		20: green20,
		30: green30,
		40: green40,
		50: green50,
		60: green60,
		70: green70,
		80: green80,
		90: green90,
		100: green100
	};
	const green10Hover = "#b6f6c8";
	const green20Hover = "#74e792";
	const green30Hover = "#36ce5e";
	const green40Hover = "#3bab5a";
	const green50Hover = "#208e3f";
	const green60Hover = "#166f31";
	const green70Hover = "#11742f";
	const green80Hover = "#05521c";
	const green90Hover = "#033b11";
	const green100Hover = "#0d300f";
	const greenHover = {
		10: green10Hover,
		20: green20Hover,
		30: green30Hover,
		40: green40Hover,
		50: green50Hover,
		60: green60Hover,
		70: green70Hover,
		80: green80Hover,
		90: green90Hover,
		100: green100Hover
	};
	const coolGray10 = "#f2f4f8";
	const coolGray20 = "#dde1e6";
	const coolGray30 = "#c1c7cd";
	const coolGray40 = "#a2a9b0";
	const coolGray50 = "#878d96";
	const coolGray60 = "#697077";
	const coolGray70 = "#4d5358";
	const coolGray80 = "#343a3f";
	const coolGray90 = "#21272a";
	const coolGray100 = "#121619";
	const coolGray = {
		10: coolGray10,
		20: coolGray20,
		30: coolGray30,
		40: coolGray40,
		50: coolGray50,
		60: coolGray60,
		70: coolGray70,
		80: coolGray80,
		90: coolGray90,
		100: coolGray100
	};
	const coolGray10Hover = "#e4e9f1";
	const coolGray20Hover = "#cdd3da";
	const coolGray30Hover = "#adb5bd";
	const coolGray40Hover = "#9199a1";
	const coolGray50Hover = "#757b85";
	const coolGray60Hover = "#585e64";
	const coolGray70Hover = "#5d646a";
	const coolGray80Hover = "#434a51";
	const coolGray90Hover = "#2b3236";
	const coolGray100Hover = "#222a2f";
	const coolGrayHover = {
		10: coolGray10Hover,
		20: coolGray20Hover,
		30: coolGray30Hover,
		40: coolGray40Hover,
		50: coolGray50Hover,
		60: coolGray60Hover,
		70: coolGray70Hover,
		80: coolGray80Hover,
		90: coolGray90Hover,
		100: coolGray100Hover
	};
	const gray10 = "#f4f4f4";
	const gray20 = "#e0e0e0";
	const gray30 = "#c6c6c6";
	const gray40 = "#a8a8a8";
	const gray50 = "#8d8d8d";
	const gray60 = "#6f6f6f";
	const gray70 = "#525252";
	const gray80 = "#393939";
	const gray90 = "#262626";
	const gray100 = "#161616";
	const gray = {
		10: gray10,
		20: gray20,
		30: gray30,
		40: gray40,
		50: gray50,
		60: gray60,
		70: gray70,
		80: gray80,
		90: gray90,
		100: gray100
	};
	const gray10Hover = "#e8e8e8";
	const gray20Hover = "#d1d1d1";
	const gray30Hover = "#b5b5b5";
	const gray40Hover = "#999999";
	const gray50Hover = "#7a7a7a";
	const gray60Hover = "#5e5e5e";
	const gray70Hover = "#636363";
	const gray80Hover = "#474747";
	const gray90Hover = "#333333";
	const gray100Hover = "#292929";
	const grayHover = {
		10: gray10Hover,
		20: gray20Hover,
		30: gray30Hover,
		40: gray40Hover,
		50: gray50Hover,
		60: gray60Hover,
		70: gray70Hover,
		80: gray80Hover,
		90: gray90Hover,
		100: gray100Hover
	};
	const warmGray10 = "#f7f3f2";
	const warmGray20 = "#e5e0df";
	const warmGray30 = "#cac5c4";
	const warmGray40 = "#ada8a8";
	const warmGray50 = "#8f8b8b";
	const warmGray60 = "#726e6e";
	const warmGray70 = "#565151";
	const warmGray80 = "#3c3838";
	const warmGray90 = "#272525";
	const warmGray100 = "#171414";
	const warmGray = {
		10: warmGray10,
		20: warmGray20,
		30: warmGray30,
		40: warmGray40,
		50: warmGray50,
		60: warmGray60,
		70: warmGray70,
		80: warmGray80,
		90: warmGray90,
		100: warmGray100
	};
	const warmGray10Hover = "#f0e8e6";
	const warmGray20Hover = "#d8d0cf";
	const warmGray30Hover = "#b9b3b1";
	const warmGray40Hover = "#9c9696";
	const warmGray50Hover = "#7f7b7b";
	const warmGray60Hover = "#605d5d";
	const warmGray70Hover = "#696363";
	const warmGray80Hover = "#4c4848";
	const warmGray90Hover = "#343232";
	const warmGray100Hover = "#2c2626";
	const warmGrayHover = {
		10: warmGray10Hover,
		20: warmGray20Hover,
		30: warmGray30Hover,
		40: warmGray40Hover,
		50: warmGray50Hover,
		60: warmGray60Hover,
		70: warmGray70Hover,
		80: warmGray80Hover,
		90: warmGray90Hover,
		100: warmGray100Hover
	};
	const colors = {
		black: { 100: black },
		blue,
		coolGray,
		cyan,
		gray,
		green,
		magenta,
		orange,
		purple,
		red,
		teal,
		warmGray,
		white: { 0: white },
		yellow
	};
	const hoverColors = {
		whiteHover,
		blackHover,
		blueHover,
		coolGrayHover,
		cyanHover,
		grayHover,
		greenHover,
		magentaHover,
		orangeHover,
		purpleHover,
		redHover,
		tealHover,
		warmGrayHover,
		yellowHover
	};
	//#endregion
	//#region src/rgb.ts
	/**
	* Copyright IBM Corp. 2018, 2026
	*
	* This source code is licensed under the Apache-2.0 license found in the
	* LICENSE file in the root directory of this source tree.
	*/
	/**
	* Parses a hex code string into an RGBA color string with the given opacity.
	*/
	const rgba = (hexCode, opacity) => {
		const values = [
			hexCode.substring(1, 3),
			hexCode.substring(3, 5),
			hexCode.substring(5, 7)
		].map((string) => parseInt(string, 16));
		return `rgba(${values[0]}, ${values[1]}, ${values[2]}, ${opacity})`;
	};
	//#endregion
	exports.black = black;
	exports.black100 = black100;
	exports.blackHover = blackHover;
	exports.blue = blue;
	exports.blue10 = blue10;
	exports.blue100 = blue100;
	exports.blue100Hover = blue100Hover;
	exports.blue10Hover = blue10Hover;
	exports.blue20 = blue20;
	exports.blue20Hover = blue20Hover;
	exports.blue30 = blue30;
	exports.blue30Hover = blue30Hover;
	exports.blue40 = blue40;
	exports.blue40Hover = blue40Hover;
	exports.blue50 = blue50;
	exports.blue50Hover = blue50Hover;
	exports.blue60 = blue60;
	exports.blue60Hover = blue60Hover;
	exports.blue70 = blue70;
	exports.blue70Hover = blue70Hover;
	exports.blue80 = blue80;
	exports.blue80Hover = blue80Hover;
	exports.blue90 = blue90;
	exports.blue90Hover = blue90Hover;
	exports.blueHover = blueHover;
	exports.colors = colors;
	exports.coolGray = coolGray;
	exports.coolGray10 = coolGray10;
	exports.coolGray100 = coolGray100;
	exports.coolGray100Hover = coolGray100Hover;
	exports.coolGray10Hover = coolGray10Hover;
	exports.coolGray20 = coolGray20;
	exports.coolGray20Hover = coolGray20Hover;
	exports.coolGray30 = coolGray30;
	exports.coolGray30Hover = coolGray30Hover;
	exports.coolGray40 = coolGray40;
	exports.coolGray40Hover = coolGray40Hover;
	exports.coolGray50 = coolGray50;
	exports.coolGray50Hover = coolGray50Hover;
	exports.coolGray60 = coolGray60;
	exports.coolGray60Hover = coolGray60Hover;
	exports.coolGray70 = coolGray70;
	exports.coolGray70Hover = coolGray70Hover;
	exports.coolGray80 = coolGray80;
	exports.coolGray80Hover = coolGray80Hover;
	exports.coolGray90 = coolGray90;
	exports.coolGray90Hover = coolGray90Hover;
	exports.coolGrayHover = coolGrayHover;
	exports.cyan = cyan;
	exports.cyan10 = cyan10;
	exports.cyan100 = cyan100;
	exports.cyan100Hover = cyan100Hover;
	exports.cyan10Hover = cyan10Hover;
	exports.cyan20 = cyan20;
	exports.cyan20Hover = cyan20Hover;
	exports.cyan30 = cyan30;
	exports.cyan30Hover = cyan30Hover;
	exports.cyan40 = cyan40;
	exports.cyan40Hover = cyan40Hover;
	exports.cyan50 = cyan50;
	exports.cyan50Hover = cyan50Hover;
	exports.cyan60 = cyan60;
	exports.cyan60Hover = cyan60Hover;
	exports.cyan70 = cyan70;
	exports.cyan70Hover = cyan70Hover;
	exports.cyan80 = cyan80;
	exports.cyan80Hover = cyan80Hover;
	exports.cyan90 = cyan90;
	exports.cyan90Hover = cyan90Hover;
	exports.cyanHover = cyanHover;
	exports.gray = gray;
	exports.gray10 = gray10;
	exports.gray100 = gray100;
	exports.gray100Hover = gray100Hover;
	exports.gray10Hover = gray10Hover;
	exports.gray20 = gray20;
	exports.gray20Hover = gray20Hover;
	exports.gray30 = gray30;
	exports.gray30Hover = gray30Hover;
	exports.gray40 = gray40;
	exports.gray40Hover = gray40Hover;
	exports.gray50 = gray50;
	exports.gray50Hover = gray50Hover;
	exports.gray60 = gray60;
	exports.gray60Hover = gray60Hover;
	exports.gray70 = gray70;
	exports.gray70Hover = gray70Hover;
	exports.gray80 = gray80;
	exports.gray80Hover = gray80Hover;
	exports.gray90 = gray90;
	exports.gray90Hover = gray90Hover;
	exports.grayHover = grayHover;
	exports.green = green;
	exports.green10 = green10;
	exports.green100 = green100;
	exports.green100Hover = green100Hover;
	exports.green10Hover = green10Hover;
	exports.green20 = green20;
	exports.green20Hover = green20Hover;
	exports.green30 = green30;
	exports.green30Hover = green30Hover;
	exports.green40 = green40;
	exports.green40Hover = green40Hover;
	exports.green50 = green50;
	exports.green50Hover = green50Hover;
	exports.green60 = green60;
	exports.green60Hover = green60Hover;
	exports.green70 = green70;
	exports.green70Hover = green70Hover;
	exports.green80 = green80;
	exports.green80Hover = green80Hover;
	exports.green90 = green90;
	exports.green90Hover = green90Hover;
	exports.greenHover = greenHover;
	exports.hoverColors = hoverColors;
	exports.magenta = magenta;
	exports.magenta10 = magenta10;
	exports.magenta100 = magenta100;
	exports.magenta100Hover = magenta100Hover;
	exports.magenta10Hover = magenta10Hover;
	exports.magenta20 = magenta20;
	exports.magenta20Hover = magenta20Hover;
	exports.magenta30 = magenta30;
	exports.magenta30Hover = magenta30Hover;
	exports.magenta40 = magenta40;
	exports.magenta40Hover = magenta40Hover;
	exports.magenta50 = magenta50;
	exports.magenta50Hover = magenta50Hover;
	exports.magenta60 = magenta60;
	exports.magenta60Hover = magenta60Hover;
	exports.magenta70 = magenta70;
	exports.magenta70Hover = magenta70Hover;
	exports.magenta80 = magenta80;
	exports.magenta80Hover = magenta80Hover;
	exports.magenta90 = magenta90;
	exports.magenta90Hover = magenta90Hover;
	exports.magentaHover = magentaHover;
	exports.orange = orange;
	exports.orange10 = orange10;
	exports.orange100 = orange100;
	exports.orange100Hover = orange100Hover;
	exports.orange10Hover = orange10Hover;
	exports.orange20 = orange20;
	exports.orange20Hover = orange20Hover;
	exports.orange30 = orange30;
	exports.orange30Hover = orange30Hover;
	exports.orange40 = orange40;
	exports.orange40Hover = orange40Hover;
	exports.orange50 = orange50;
	exports.orange50Hover = orange50Hover;
	exports.orange60 = orange60;
	exports.orange60Hover = orange60Hover;
	exports.orange70 = orange70;
	exports.orange70Hover = orange70Hover;
	exports.orange80 = orange80;
	exports.orange80Hover = orange80Hover;
	exports.orange90 = orange90;
	exports.orange90Hover = orange90Hover;
	exports.orangeHover = orangeHover;
	exports.purple = purple;
	exports.purple10 = purple10;
	exports.purple100 = purple100;
	exports.purple100Hover = purple100Hover;
	exports.purple10Hover = purple10Hover;
	exports.purple20 = purple20;
	exports.purple20Hover = purple20Hover;
	exports.purple30 = purple30;
	exports.purple30Hover = purple30Hover;
	exports.purple40 = purple40;
	exports.purple40Hover = purple40Hover;
	exports.purple50 = purple50;
	exports.purple50Hover = purple50Hover;
	exports.purple60 = purple60;
	exports.purple60Hover = purple60Hover;
	exports.purple70 = purple70;
	exports.purple70Hover = purple70Hover;
	exports.purple80 = purple80;
	exports.purple80Hover = purple80Hover;
	exports.purple90 = purple90;
	exports.purple90Hover = purple90Hover;
	exports.purpleHover = purpleHover;
	exports.red = red;
	exports.red10 = red10;
	exports.red100 = red100;
	exports.red100Hover = red100Hover;
	exports.red10Hover = red10Hover;
	exports.red20 = red20;
	exports.red20Hover = red20Hover;
	exports.red30 = red30;
	exports.red30Hover = red30Hover;
	exports.red40 = red40;
	exports.red40Hover = red40Hover;
	exports.red50 = red50;
	exports.red50Hover = red50Hover;
	exports.red60 = red60;
	exports.red60Hover = red60Hover;
	exports.red70 = red70;
	exports.red70Hover = red70Hover;
	exports.red80 = red80;
	exports.red80Hover = red80Hover;
	exports.red90 = red90;
	exports.red90Hover = red90Hover;
	exports.redHover = redHover;
	exports.rgba = rgba;
	exports.teal = teal;
	exports.teal10 = teal10;
	exports.teal100 = teal100;
	exports.teal100Hover = teal100Hover;
	exports.teal10Hover = teal10Hover;
	exports.teal20 = teal20;
	exports.teal20Hover = teal20Hover;
	exports.teal30 = teal30;
	exports.teal30Hover = teal30Hover;
	exports.teal40 = teal40;
	exports.teal40Hover = teal40Hover;
	exports.teal50 = teal50;
	exports.teal50Hover = teal50Hover;
	exports.teal60 = teal60;
	exports.teal60Hover = teal60Hover;
	exports.teal70 = teal70;
	exports.teal70Hover = teal70Hover;
	exports.teal80 = teal80;
	exports.teal80Hover = teal80Hover;
	exports.teal90 = teal90;
	exports.teal90Hover = teal90Hover;
	exports.tealHover = tealHover;
	exports.warmGray = warmGray;
	exports.warmGray10 = warmGray10;
	exports.warmGray100 = warmGray100;
	exports.warmGray100Hover = warmGray100Hover;
	exports.warmGray10Hover = warmGray10Hover;
	exports.warmGray20 = warmGray20;
	exports.warmGray20Hover = warmGray20Hover;
	exports.warmGray30 = warmGray30;
	exports.warmGray30Hover = warmGray30Hover;
	exports.warmGray40 = warmGray40;
	exports.warmGray40Hover = warmGray40Hover;
	exports.warmGray50 = warmGray50;
	exports.warmGray50Hover = warmGray50Hover;
	exports.warmGray60 = warmGray60;
	exports.warmGray60Hover = warmGray60Hover;
	exports.warmGray70 = warmGray70;
	exports.warmGray70Hover = warmGray70Hover;
	exports.warmGray80 = warmGray80;
	exports.warmGray80Hover = warmGray80Hover;
	exports.warmGray90 = warmGray90;
	exports.warmGray90Hover = warmGray90Hover;
	exports.warmGrayHover = warmGrayHover;
	exports.white = white;
	exports.white0 = white0;
	exports.whiteHover = whiteHover;
	exports.yellow = yellow;
	exports.yellow10 = yellow10;
	exports.yellow100 = yellow100;
	exports.yellow100Hover = yellow100Hover;
	exports.yellow10Hover = yellow10Hover;
	exports.yellow20 = yellow20;
	exports.yellow20Hover = yellow20Hover;
	exports.yellow30 = yellow30;
	exports.yellow30Hover = yellow30Hover;
	exports.yellow40 = yellow40;
	exports.yellow40Hover = yellow40Hover;
	exports.yellow50 = yellow50;
	exports.yellow50Hover = yellow50Hover;
	exports.yellow60 = yellow60;
	exports.yellow60Hover = yellow60Hover;
	exports.yellow70 = yellow70;
	exports.yellow70Hover = yellow70Hover;
	exports.yellow80 = yellow80;
	exports.yellow80Hover = yellow80Hover;
	exports.yellow90 = yellow90;
	exports.yellow90Hover = yellow90Hover;
	exports.yellowHover = yellowHover;
});
