// @carbon/react/lib/components/FluidTextArea/FluidTextArea.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

const require_runtime = require("../../_virtual/_rolldown/runtime.js");
const require_usePrefix = require("../../internal/usePrefix.js");
const require_deprecate = require("../../prop-types/deprecate.js");
const require_FormContext = require("../FluidForm/FormContext.js");
const require_index = require("../TextArea/index.js");
let classnames = require("classnames");
classnames = require_runtime.__toESM(classnames);
let react = require("react");
react = require_runtime.__toESM(react);
let prop_types = require("prop-types");
prop_types = require_runtime.__toESM(prop_types);
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/components/FluidTextArea/FluidTextArea.tsx
/**
* Copyright IBM Corp. 2016, 2025
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
const FluidTextArea = ({ className, ...other }) => {
	const classNames = (0, classnames.default)(`${require_usePrefix.usePrefix()}--text-area--fluid`, className);
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_FormContext.FormContext.Provider, {
		value: { isFluid: true },
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_index.default, {
			className: classNames,
			...other
		})
	});
};
FluidTextArea.propTypes = {
	className: prop_types.default.string,
	cols: prop_types.default.number,
	defaultValue: prop_types.default.oneOfType([prop_types.default.string, prop_types.default.number]),
	disabled: prop_types.default.bool,
	enableCounter: prop_types.default.bool,
	helperText: prop_types.default.node,
	hideLabel: prop_types.default.bool,
	id: prop_types.default.string,
	invalid: prop_types.default.bool,
	invalidText: prop_types.default.node,
	labelText: prop_types.default.node.isRequired,
	light: require_deprecate.deprecate(prop_types.default.bool, "The `light` prop for `FluidTextArea` has been deprecated in favor of the new `Layer` component. It will be removed in the next major release."),
	maxCount: prop_types.default.number,
	onChange: prop_types.default.func,
	onClick: prop_types.default.func,
	placeholder: prop_types.default.string,
	rows: prop_types.default.number,
	value: prop_types.default.oneOfType([prop_types.default.string, prop_types.default.number]),
	warn: prop_types.default.bool,
	warnText: prop_types.default.node,
	readOnly: prop_types.default.bool
};
//#endregion
exports.default = FluidTextArea;
