// @carbon/react/lib/components/AILabel/index.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

const require_runtime = require("../../_virtual/_rolldown/runtime.js");
const require_usePrefix = require("../../internal/usePrefix.js");
const require_useId = require("../../internal/useId.js");
const require_deprecate = require("../../prop-types/deprecate.js");
const require_index = require("../IconButton/index.js");
const require_index$1 = require("../Toggletip/index.js");
let classnames = require("classnames");
classnames = require_runtime.__toESM(classnames);
let react = require("react");
react = require_runtime.__toESM(react);
let prop_types = require("prop-types");
prop_types = require_runtime.__toESM(prop_types);
let react_jsx_runtime = require("react/jsx-runtime");
let _carbon_icons_react = require("@carbon/icons-react");
//#region src/components/AILabel/index.tsx
/**
* Copyright IBM Corp. 2016, 2026
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
const AILabelContent = ({ className, children }) => {
	const prefix = require_usePrefix.usePrefix();
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_index$1.ToggletipContent, {
		className: (0, classnames.default)(className, {
			[`${prefix}--ai-label-content`]: true,
			[`${prefix}--ai-label-content--with-actions`]: false
		}),
		children
	});
};
AILabelContent.displayName = "AILabelContent";
AILabelContent.propTypes = {
	children: prop_types.default.node,
	className: prop_types.default.string
};
const AILabelActions = ({ className, children }) => {
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_index$1.ToggletipActions, {
		className: (0, classnames.default)(className, { [`${require_usePrefix.usePrefix()}--ai-label-actions`]: true }),
		children
	});
};
AILabelActions.displayName = "AILabelActions";
AILabelActions.propTypes = {
	children: prop_types.default.node,
	className: prop_types.default.string
};
const AILabel = react.default.forwardRef(function AILabel({ aiText = "AI", aiTextLabel, textLabel, align, autoAlign = true, children, className, kind = "default", onRevertClick, revertActive, revertLabel = "Revert to AI input", slugLabel = "Show information", ["aria-label"]: ariaLabel = "Show information", size = "xs", ...rest }, ref) {
	const prefix = require_usePrefix.usePrefix();
	const id = require_useId.useId("AILabel");
	const aiLabelClasses = (0, classnames.default)(className, {
		[`${prefix}--ai-label`]: true,
		[`${prefix}--ai-label--revert`]: revertActive
	});
	const aiLabelButtonClasses = (0, classnames.default)({
		[`${prefix}--ai-label__button`]: true,
		[`${prefix}--ai-label__button--${size}`]: size,
		[`${prefix}--ai-label__button--${kind}`]: kind,
		[`${prefix}--ai-label__button--inline-with-content`]: kind === "inline" && (aiTextLabel || textLabel)
	});
	const handleOnRevertClick = (evt) => {
		if (onRevertClick) onRevertClick(evt);
	};
	const ariaLabelText = !aiTextLabel && !textLabel ? `${aiText} ${slugLabel || ariaLabel}` : `${aiText} ${aiTextLabel || textLabel}`;
	const isSmallIcon = [
		"xs",
		"2xs",
		"mini"
	].includes(size);
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		className: aiLabelClasses,
		ref,
		id,
		children: revertActive ? /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_index.IconButton, {
			onClick: handleOnRevertClick,
			kind: "ghost",
			size: "sm",
			label: revertLabel,
			...rest,
			children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_carbon_icons_react.Undo, {})
		}) : /* @__PURE__ */ (0, react_jsx_runtime.jsxs)(require_index$1.default, {
			align,
			autoAlign,
			alignmentAxisOffset: isSmallIcon ? -24 : 0,
			...rest,
			children: [/* @__PURE__ */ (0, react_jsx_runtime.jsxs)(require_index$1.ToggletipButton, {
				className: aiLabelButtonClasses,
				label: kind === "inline" ? "" : ariaLabelText,
				children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", {
					className: `${prefix}--ai-label__text`,
					children: aiText
				}), kind === "inline" && (aiTextLabel || textLabel) && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", {
					className: `${prefix}--ai-label__additional-text`,
					children: aiTextLabel || textLabel
				})]
			}), children]
		})
	});
});
AILabel.displayName = "AILabel";
AILabel.propTypes = {
	...require_index$1.default.propTypes,
	AILabelContent: prop_types.default.node,
	aiText: prop_types.default.string,
	aiTextLabel: require_deprecate.deprecate(prop_types.default.string, "`aiTextLabel` on `AILabel` has been deprecated - Please use the `textLabel` prop instead"),
	"aria-label": prop_types.default.string,
	kind: prop_types.default.oneOf(["default", "inline"]),
	onRevertClick: prop_types.default.func,
	revertActive: prop_types.default.bool,
	revertLabel: prop_types.default.string,
	size: prop_types.default.oneOf([
		"mini",
		"2xs",
		"xs",
		"sm",
		"md",
		"lg",
		"xl"
	]),
	slugLabel: require_deprecate.deprecate(prop_types.default.string, "`slugLabel` on `AILabel` has been deprecated - Please use the `ariaLabel` prop instead"),
	textLabel: prop_types.default.string
};
//#endregion
exports.AILabel = AILabel;
exports.AILabelActions = AILabelActions;
exports.AILabelContent = AILabelContent;
