// @carbon/react/lib/components/FluidSearch/FluidSearch.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

const require_runtime = require("../../_virtual/_rolldown/runtime.js");
const require_usePrefix = require("../../internal/usePrefix.js");
const require_FormContext = require("../FluidForm/FormContext.js");
const require_index = require("../Search/index.js");
let classnames = require("classnames");
classnames = require_runtime.__toESM(classnames);
let react = require("react");
react = require_runtime.__toESM(react);
let prop_types = require("prop-types");
prop_types = require_runtime.__toESM(prop_types);
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/components/FluidSearch/FluidSearch.tsx
/**
* Copyright IBM Corp. 2022, 2025
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
const FluidSearch = (0, react.forwardRef)(({ className, ...other }, ref) => {
	const classNames = (0, classnames.default)(`${require_usePrefix.usePrefix()}--search--fluid`, className);
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_FormContext.FormContext.Provider, {
		value: { isFluid: true },
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_index.default, {
			ref,
			className: classNames,
			...other
		})
	});
});
FluidSearch.propTypes = {
	autoComplete: prop_types.default.string,
	className: prop_types.default.string,
	closeButtonLabelText: prop_types.default.string,
	defaultValue: prop_types.default.oneOfType([prop_types.default.string, prop_types.default.number]),
	disabled: prop_types.default.bool,
	id: prop_types.default.string,
	labelText: prop_types.default.node.isRequired,
	onChange: prop_types.default.func,
	onClear: prop_types.default.func,
	onKeyDown: prop_types.default.func,
	placeholder: prop_types.default.string,
	role: prop_types.default.string,
	type: prop_types.default.string,
	value: prop_types.default.oneOfType([prop_types.default.string, prop_types.default.number])
};
//#endregion
exports.default = FluidSearch;
