// @carbon/react/es/components/PageHeader/index.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { __exportAll } from "../../_virtual/_rolldown/runtime.js";
import { BreadcrumbBar, Content, ContentPageActions, ContentText, HeroImage, PageHeader, PageHeaderBreadcrumbBar, PageHeaderContent, PageHeaderContentPageActions, PageHeaderContentText, PageHeaderHeroImage, PageHeaderTabBar, Root, TabBar } from "./PageHeader.js";
//#region src/components/PageHeader/index.tsx
var PageHeader_exports = /* @__PURE__ */ __exportAll({
	BreadcrumbBar: () => BreadcrumbBar,
	Content: () => Content,
	ContentPageActions: () => ContentPageActions,
	ContentText: () => ContentText,
	HeroImage: () => HeroImage,
	PageHeader: () => PageHeader,
	PageHeaderBreadcrumbBar: () => PageHeaderBreadcrumbBar,
	PageHeaderContent: () => PageHeaderContent,
	PageHeaderContentPageActions: () => PageHeaderContentPageActions,
	PageHeaderContentText: () => PageHeaderContentText,
	PageHeaderHeroImage: () => PageHeaderHeroImage,
	PageHeaderTabBar: () => PageHeaderTabBar,
	Root: () => Root,
	TabBar: () => TabBar
});
//#endregion
export { PageHeader_exports };
