// @carbon/react/lib/components/Menu/MenuContext.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

require("../../_virtual/_rolldown/runtime.js");
let react = require("react");
//#region src/components/Menu/MenuContext.ts
/**
* Copyright IBM Corp. 2023, 2026
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
const menuDefaultState = {
	isRoot: true,
	hasIcons: false,
	hasSelectableItems: false,
	size: null,
	items: [],
	requestCloseRoot: () => {}
};
function menuReducer(state, action) {
	switch (action.type) {
		case "enableIcons": return {
			...state,
			hasIcons: true
		};
		case "enableSelectableItems": return {
			...state,
			hasSelectableItems: true
		};
		case "registerItem": {
			const newItem = action.payload;
			const items = state.items.filter((item) => item.ref.current);
			const next = newItem.ref.current?.nextElementSibling;
			const idx = items.findIndex((item) => item.ref.current === next);
			items.splice(idx < 0 ? items.length : idx, 0, newItem);
			return {
				...state,
				items
			};
		}
	}
}
const MenuContext = (0, react.createContext)({
	state: menuDefaultState,
	dispatch: () => {}
});
//#endregion
exports.MenuContext = MenuContext;
exports.menuReducer = menuReducer;
