// @carbon/react/es/components/FormLabel/FormLabel.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { usePrefix } from "../../internal/usePrefix.js";
import { Text } from "../Text/Text.js";
import classNames from "classnames";
import "react";
import PropTypes from "prop-types";
import { jsx } from "react/jsx-runtime";
//#region src/components/FormLabel/FormLabel.tsx
/**
* Copyright IBM Corp. 2016, 2023
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
function FormLabel({ className: customClassName, children, id, ...rest }) {
	const prefix = usePrefix();
	return /* @__PURE__ */ jsx(Text, {
		as: "label",
		htmlFor: id,
		className: classNames(`${prefix}--label`, `${prefix}--label--no-margin`, customClassName),
		...rest,
		children
	});
}
FormLabel.propTypes = {
	children: PropTypes.node,
	className: PropTypes.string,
	id: PropTypes.string
};
//#endregion
export { FormLabel as default };
