// @carbon/react/lib/components/FluidTimePickerSelect/FluidTimePickerSelect.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

const require_runtime = require("../../_virtual/_rolldown/runtime.js");
const require_index = require("../FluidSelect/index.js");
let react = require("react");
react = require_runtime.__toESM(react);
let prop_types = require("prop-types");
prop_types = require_runtime.__toESM(prop_types);
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/components/FluidTimePickerSelect/FluidTimePickerSelect.tsx
/**
* Copyright IBM Corp. 2022, 2026
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
const FluidTimePickerSelect = react.default.forwardRef(({ children, className, ...other }, ref) => {
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_index.default, {
		className,
		ref,
		...other,
		children
	});
});
FluidTimePickerSelect.propTypes = {
	children: prop_types.default.node,
	className: prop_types.default.string,
	defaultValue: prop_types.default.any,
	disabled: prop_types.default.bool,
	id: prop_types.default.string.isRequired,
	labelText: prop_types.default.node,
	onChange: prop_types.default.func,
	readOnly: prop_types.default.bool
};
//#endregion
exports.default = FluidTimePickerSelect;
