// @carbon/react/lib/components/Checkbox/Checkbox.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

const require_runtime = require("../../_virtual/_rolldown/runtime.js");
const require_usePrefix = require("../../internal/usePrefix.js");
const require_Text = require("../Text/Text.js");
const require_useId = require("../../internal/useId.js");
const require_noopFn = require("../../internal/noopFn.js");
const require_deprecate = require("../../prop-types/deprecate.js");
const require_utils = require("../../internal/utils.js");
const require_hasHelperText = require("../../internal/hasHelperText.js");
const require_useNormalizedInputProps = require("../../internal/useNormalizedInputProps.js");
const require_index = require("../AILabel/index.js");
let classnames = require("classnames");
classnames = require_runtime.__toESM(classnames);
let react = require("react");
react = require_runtime.__toESM(react);
let prop_types = require("prop-types");
prop_types = require_runtime.__toESM(prop_types);
let react_jsx_runtime = require("react/jsx-runtime");
let _carbon_icons_react = require("@carbon/icons-react");
//#region src/components/Checkbox/Checkbox.tsx
/**
* Copyright IBM Corp. 2016, 2026
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
const Checkbox = react.default.forwardRef(({ className, decorator, helperText, id, labelText, onChange = require_noopFn.noopFn, onClick, indeterminate = false, invalid, invalidText, hideLabel, readOnly, title = "", warn, warnText, disabled, slug, ...other }, ref) => {
	const prefix = require_usePrefix.usePrefix();
	const normalizedProps = require_useNormalizedInputProps.useNormalizedInputProps({
		id,
		readOnly,
		disabled: disabled ?? false,
		invalid: invalid ?? false,
		invalidText,
		warn: warn ?? false,
		warnText
	});
	const showWarning = normalizedProps.warn;
	const showHelper = !normalizedProps.invalid && !normalizedProps.warn;
	const checkboxGroupInstanceId = require_useId.useId();
	const hasHelper = require_hasHelperText.hasHelperText(helperText);
	const helperId = !hasHelper ? void 0 : `checkbox-helper-text-${checkboxGroupInstanceId}`;
	const helper = hasHelper && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		id: helperId,
		className: `${prefix}--form__helper-text`,
		children: helperText
	});
	const wrapperClasses = (0, classnames.default)(`${prefix}--form-item`, `${prefix}--checkbox-wrapper`, className, {
		[`${prefix}--checkbox-wrapper--readonly`]: readOnly,
		[`${prefix}--checkbox-wrapper--invalid`]: normalizedProps.invalid,
		[`${prefix}--checkbox-wrapper--warning`]: showWarning,
		[`${prefix}--checkbox-wrapper--slug`]: slug,
		[`${prefix}--checkbox-wrapper--decorator`]: decorator
	});
	const innerLabelClasses = (0, classnames.default)(`${prefix}--checkbox-label-text`, { [`${prefix}--visually-hidden`]: hideLabel });
	const candidate = slug ?? decorator;
	const normalizedDecorator = require_utils.isComponentElement(candidate, require_index.AILabel) ? (0, react.cloneElement)(candidate, { size: candidate.props.kind === "inline" ? "md" : "mini" }) : candidate;
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
		className: wrapperClasses,
		children: [
			/* @__PURE__ */ (0, react_jsx_runtime.jsx)("input", {
				...other,
				disabled,
				type: "checkbox",
				"data-invalid": normalizedProps.invalid ? true : void 0,
				onChange: (evt) => {
					if (!readOnly && onChange) onChange(evt, {
						checked: evt.target.checked,
						id
					});
				},
				className: `${prefix}--checkbox`,
				id,
				ref: (el) => {
					if (el) el.indeterminate = indeterminate ?? false;
					if (typeof ref === "function") ref(el);
					else if (ref && "current" in ref) ref.current = el;
				},
				"aria-readonly": readOnly,
				onClick: (evt) => {
					if (readOnly) evt.preventDefault();
					if (onClick) onClick(evt);
				}
			}),
			/* @__PURE__ */ (0, react_jsx_runtime.jsx)("label", {
				htmlFor: id,
				className: `${prefix}--checkbox-label`,
				title,
				children: /* @__PURE__ */ (0, react_jsx_runtime.jsxs)(require_Text.Text, {
					as: "div",
					className: innerLabelClasses,
					children: [labelText, slug ? normalizedDecorator : decorator ? /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
						className: `${prefix}--checkbox-wrapper-inner--decorator`,
						children: normalizedDecorator
					}) : ""]
				})
			}),
			/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
				className: `${prefix}--checkbox__validation-msg`,
				children: [normalizedProps.invalid && /* @__PURE__ */ (0, react_jsx_runtime.jsxs)(react_jsx_runtime.Fragment, { children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)(_carbon_icons_react.WarningFilled, { className: `${prefix}--checkbox__invalid-icon` }), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
					className: `${prefix}--form-requirement`,
					children: invalidText
				})] }), showWarning && /* @__PURE__ */ (0, react_jsx_runtime.jsxs)(react_jsx_runtime.Fragment, { children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)(_carbon_icons_react.WarningAltFilled, { className: `${prefix}--checkbox__invalid-icon ${prefix}--checkbox__invalid-icon--warning` }), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
					className: `${prefix}--form-requirement`,
					children: warnText
				})] })]
			}),
			showHelper && helper
		]
	});
});
Checkbox.propTypes = {
	checked: prop_types.default.bool,
	className: prop_types.default.string,
	decorator: prop_types.default.node,
	defaultChecked: prop_types.default.bool,
	disabled: prop_types.default.bool,
	helperText: prop_types.default.node,
	hideLabel: prop_types.default.bool,
	id: prop_types.default.string.isRequired,
	indeterminate: prop_types.default.bool,
	invalid: prop_types.default.bool,
	invalidText: prop_types.default.node,
	labelText: prop_types.default.node.isRequired,
	onChange: prop_types.default.func,
	readOnly: prop_types.default.bool,
	slug: require_deprecate.deprecate(prop_types.default.node, "The `slug` prop has been deprecated and will be removed in the next major version. Use the decorator prop instead."),
	title: prop_types.default.string,
	warn: prop_types.default.bool,
	warnText: prop_types.default.node
};
Checkbox.displayName = "Checkbox";
//#endregion
exports.default = Checkbox;
