// @carbon/react/lib/components/UIShell/HeaderMenuButton.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

const require_runtime = require("../../_virtual/_rolldown/runtime.js");
const require_usePrefix = require("../../internal/usePrefix.js");
const require_AriaPropTypes = require("../../prop-types/AriaPropTypes.js");
let classnames = require("classnames");
classnames = require_runtime.__toESM(classnames);
let react = require("react");
react = require_runtime.__toESM(react);
let prop_types = require("prop-types");
prop_types = require_runtime.__toESM(prop_types);
let react_jsx_runtime = require("react/jsx-runtime");
let _carbon_icons_react = require("@carbon/icons-react");
//#region src/components/UIShell/HeaderMenuButton.tsx
/**
* Copyright IBM Corp. 2016, 2023
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
function HeaderMenuButton({ "aria-label": ariaLabel, "aria-labelledby": ariaLabelledBy, className: customClassName, renderMenuIcon, renderCloseIcon, isActive, isCollapsible, ...rest }) {
	const prefix = require_usePrefix.usePrefix();
	const className = (0, classnames.default)({
		...typeof customClassName === "string" && { [customClassName]: !!customClassName },
		[`${prefix}--header__action`]: true,
		[`${prefix}--header__menu-trigger`]: true,
		[`${prefix}--header__action--active`]: isActive,
		[`${prefix}--header__menu-toggle`]: true,
		[`${prefix}--header__menu-toggle__hidden`]: !isCollapsible
	});
	const menuIcon = renderMenuIcon ? renderMenuIcon : /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_carbon_icons_react.Menu, { size: 20 });
	const closeIcon = renderCloseIcon ? renderCloseIcon : /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_carbon_icons_react.Close, { size: 20 });
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("button", {
		...rest,
		"aria-label": ariaLabel,
		"aria-labelledby": ariaLabelledBy,
		className,
		title: ariaLabel,
		type: "button",
		children: isActive ? closeIcon : menuIcon
	});
}
HeaderMenuButton.propTypes = {
	...require_AriaPropTypes.AriaLabelPropType,
	className: prop_types.default.string,
	isActive: prop_types.default.bool,
	isCollapsible: prop_types.default.bool,
	onClick: prop_types.default.func
};
//#endregion
exports.default = HeaderMenuButton;
