// @carbon/react/lib/components/UIShell/HeaderMenuItem.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

const require_runtime = require("../../_virtual/_rolldown/runtime.js");
const require_usePrefix = require("../../internal/usePrefix.js");
const require_deprecate = require("../../prop-types/deprecate.js");
const require_Link = require("./Link.js");
let classnames = require("classnames");
classnames = require_runtime.__toESM(classnames);
let react = require("react");
react = require_runtime.__toESM(react);
let prop_types = require("prop-types");
prop_types = require_runtime.__toESM(prop_types);
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/components/UIShell/HeaderMenuItem.tsx
/**
* Copyright IBM Corp. 2016, 2025
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
const HeaderMenuItem = (0, react.forwardRef)(function HeaderMenuItem({ className, isActive, isCurrentPage, "aria-current": ariaCurrent, children, role, tabIndex, ...rest }, ref) {
	const prefix = require_usePrefix.usePrefix();
	const resolvedTabIndex = tabIndex ?? 0;
	if (isCurrentPage) isActive = isCurrentPage;
	const hasCurrentClass = isActive && ariaCurrent !== "page";
	const linkClassName = (0, classnames.default)({
		[`${prefix}--header__menu-item`]: true,
		[`${prefix}--header__menu-item--current`]: hasCurrentClass
	});
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("li", {
		className,
		role,
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_Link.default, {
			...rest,
			"aria-current": hasCurrentClass ? true : ariaCurrent,
			className: linkClassName,
			ref,
			tabIndex: resolvedTabIndex,
			children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)("span", {
				className: `${prefix}--text-truncate--end`,
				children
			})
		})
	});
});
HeaderMenuItem.displayName = "HeaderMenuItem";
HeaderMenuItem.propTypes = {
	...require_Link.LinkPropTypes,
	children: prop_types.default.node.isRequired,
	className: prop_types.default.string,
	isActive: prop_types.default.bool,
	isCurrentPage: require_deprecate.deprecate(prop_types.default.bool, "The `isCurrentPage` prop for `HeaderMenuItem` has been deprecated. Please use `isActive` instead. This will be removed in the next major release."),
	role: prop_types.default.string,
	tabIndex: prop_types.default.number
};
//#endregion
exports.default = HeaderMenuItem;
