// @carbon/react/lib/components/UIShell/SideNav.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

const require_runtime = require("../../_virtual/_rolldown/runtime.js");
const require_usePrefix = require("../../internal/usePrefix.js");
const require_keys = require("../../internal/keyboard/keys.js");
const require_match = require("../../internal/keyboard/match.js");
const require_useMergedRefs = require("../../internal/useMergedRefs.js");
const require_useEvent = require("../../internal/useEvent.js");
const require_useDelayedState = require("../../internal/useDelayedState.js");
const require_useMatchMedia = require("../../internal/useMatchMedia.js");
const require_AriaPropTypes = require("../../prop-types/AriaPropTypes.js");
const require_SideNavContext = require("./SideNavContext.js");
let classnames = require("classnames");
classnames = require_runtime.__toESM(classnames);
let react = require("react");
react = require_runtime.__toESM(react);
let prop_types = require("prop-types");
prop_types = require_runtime.__toESM(prop_types);
let react_jsx_runtime = require("react/jsx-runtime");
let _carbon_layout = require("@carbon/layout");
//#region src/components/UIShell/SideNav.tsx
/**
* Copyright IBM Corp. 2016, 2026
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
const SideNav = (0, react.forwardRef)((props, ref) => {
	const { expanded: expandedProp, defaultExpanded = false, isChildOfHeader = true, "aria-label": ariaLabel, "aria-labelledby": ariaLabelledBy, children, onToggle, className: customClassName, href, isFixedNav = false, isRail, isPersistent = true, addFocusListeners = true, addMouseListeners = true, onOverlayClick, onSideNavBlur, enterDelayMs = 100, ...other } = props;
	const prefix = require_usePrefix.usePrefix();
	const { current: controlled } = (0, react.useRef)(expandedProp !== void 0);
	const [expandedState, setExpandedState] = require_useDelayedState.useDelayedState(defaultExpanded);
	const [expandedViaHoverState, setExpandedViaHoverState] = require_useDelayedState.useDelayedState(defaultExpanded);
	const expanded = controlled ? expandedProp : expandedState;
	const sideNavRef = (0, react.useRef)(null);
	const navRef = require_useMergedRefs.useMergedRefs([sideNavRef, ref]);
	const handleToggle = (event, value = !expanded) => {
		if (!controlled) setExpandedState(value, enterDelayMs);
		if (onToggle) onToggle(event, value);
		if (controlled || isRail) setExpandedViaHoverState(value, enterDelayMs);
	};
	const accessibilityLabel = {
		"aria-label": ariaLabel,
		"aria-labelledby": ariaLabelledBy
	};
	const className = (0, classnames.default)(customClassName, {
		[`${prefix}--side-nav`]: true,
		[`${prefix}--side-nav--expanded`]: expanded || expandedViaHoverState,
		[`${prefix}--side-nav--collapsed`]: !expanded && isFixedNav,
		[`${prefix}--side-nav--rail`]: isRail,
		[`${prefix}--side-nav--ux`]: isChildOfHeader,
		[`${prefix}--side-nav--hidden`]: !isPersistent
	});
	const overlayClassName = (0, classnames.default)({
		[`${prefix}--side-nav__overlay`]: true,
		[`${prefix}--side-nav__overlay-active`]: expanded || expandedViaHoverState
	});
	const currentExpansionState = controlled ? expandedViaHoverState || expanded : expanded;
	const eventHandlers = {};
	if (addFocusListeners) {
		eventHandlers.onFocus = (event) => {
			if (!event.currentTarget.contains(event.relatedTarget) && isRail) handleToggle(event, true);
		};
		eventHandlers.onBlur = (event) => {
			if (!event.currentTarget.contains(event.relatedTarget)) handleToggle(event, false);
			if (!event.currentTarget.contains(event.relatedTarget) && expanded && !isFixedNav) {
				if (onSideNavBlur) onSideNavBlur();
			}
		};
		eventHandlers.onKeyDown = (event) => {
			if (require_match.match(event, require_keys.Escape)) {
				handleToggle(event, false);
				if (href) window.location.href = href;
			}
		};
	}
	if (addMouseListeners && isRail) {
		eventHandlers.onMouseEnter = () => {
			handleToggle(true, true);
		};
		eventHandlers.onMouseLeave = () => {
			setExpandedState(false);
			setExpandedViaHoverState(false);
			handleToggle(false, false);
		};
		eventHandlers.onClick = () => {
			setExpandedState(true);
			setExpandedViaHoverState(true);
			handleToggle(true, true);
		};
	}
	require_useEvent.useWindowEvent("keydown", (event) => {
		const focusedElement = document.activeElement;
		if (require_match.match(event, require_keys.Tab) && expanded && !isFixedNav && sideNavRef.current && focusedElement?.classList.contains(`${prefix}--header__menu-toggle`) && !focusedElement.closest("nav")) sideNavRef.current.focus();
	});
	const isLg = require_useMatchMedia.useMatchMedia(`(min-width: ${_carbon_layout.breakpoints.lg.width})`);
	const inertEnabled = !isRail ? !(expanded || isLg) : false;
	(0, react.useEffect)(() => {
		const node = sideNavRef.current;
		if (!node) return;
		if (inertEnabled) node.setAttribute("inert", "");
		else node.removeAttribute("inert");
	}, [inertEnabled]);
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)(require_SideNavContext.SideNavContextProvider, {
		isRail,
		isSideNavExpanded: currentExpansionState,
		children: [isFixedNav ? null : /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
			className: overlayClassName,
			onClick: onOverlayClick
		}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("nav", {
			tabIndex: -1,
			ref: navRef,
			className: `${prefix}--side-nav__navigation ${className}`,
			...accessibilityLabel,
			...eventHandlers,
			...other,
			children
		})]
	});
});
SideNav.displayName = "SideNav";
SideNav.propTypes = {
	...require_AriaPropTypes.AriaLabelPropType,
	addFocusListeners: prop_types.default.bool,
	addMouseListeners: prop_types.default.bool,
	className: prop_types.default.string,
	defaultExpanded: prop_types.default.bool,
	enterDelayMs: prop_types.default.number,
	expanded: prop_types.default.bool,
	href: prop_types.default.string,
	isChildOfHeader: prop_types.default.bool,
	isFixedNav: prop_types.default.bool,
	isPersistent: prop_types.default.bool,
	isRail: prop_types.default.bool,
	onOverlayClick: prop_types.default.func,
	onSideNavBlur: prop_types.default.func,
	onToggle: prop_types.default.func
};
//#endregion
exports.default = SideNav;
