// @carbon/react/lib/components/UIShell/SideNavLink.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

const require_runtime = require("../../_virtual/_rolldown/runtime.js");
const require_usePrefix = require("../../internal/usePrefix.js");
const require_Link = require("./Link.js");
const require_SideNavContext = require("./SideNavContext.js");
const require_SideNavIcon = require("./SideNavIcon.js");
const require_SideNavItem = require("./SideNavItem.js");
const require_SideNavLinkText = require("./SideNavLinkText.js");
let classnames = require("classnames");
classnames = require_runtime.__toESM(classnames);
let react = require("react");
react = require_runtime.__toESM(react);
let prop_types = require("prop-types");
prop_types = require_runtime.__toESM(prop_types);
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/components/UIShell/SideNavLink.tsx
/**
* Copyright IBM Corp. 2016, 2026
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
const SideNavLink = (0, react.forwardRef)((props, ref) => {
	const { children, className: customClassName, renderIcon: IconElement, isActive, isSideNavExpanded, large = false, tabIndex, ...rest } = props;
	const { isRail, isSideNavExpanded: contextIsSideNavExpanded } = (0, react.useContext)(require_SideNavContext.SideNavContext);
	const currentIsSideNavExpanded = isSideNavExpanded ?? contextIsSideNavExpanded;
	const prefix = require_usePrefix.usePrefix();
	const className = (0, classnames.default)({
		[`${prefix}--side-nav__link`]: true,
		[`${prefix}--side-nav__link--current`]: isActive,
		[customClassName]: !!customClassName
	});
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_SideNavItem.default, {
		large,
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsxs)(require_Link.default, {
			...rest,
			className,
			ref,
			tabIndex: tabIndex === void 0 ? !currentIsSideNavExpanded && !isRail ? -1 : 0 : tabIndex,
			children: [IconElement && /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_SideNavIcon.default, {
				small: true,
				children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(IconElement, {})
			}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_SideNavLinkText.default, { children })]
		})
	});
});
SideNavLink.displayName = "SideNavLink";
SideNavLink.propTypes = {
	...require_Link.LinkPropTypes,
	children: prop_types.default.node.isRequired,
	className: prop_types.default.string,
	isActive: prop_types.default.bool,
	isSideNavExpanded: prop_types.default.bool,
	large: prop_types.default.bool,
	renderIcon: prop_types.default.oneOfType([prop_types.default.func, prop_types.default.object]),
	tabIndex: prop_types.default.number
};
//#endregion
exports.default = SideNavLink;
