// @carbon/react/lib/components/UIShell/SideNavSwitcher.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

const require_runtime = require("../../_virtual/_rolldown/runtime.js");
const require_usePrefix = require("../../internal/usePrefix.js");
const require_useId = require("../../internal/useId.js");
let classnames = require("classnames");
classnames = require_runtime.__toESM(classnames);
let react = require("react");
react = require_runtime.__toESM(react);
let prop_types = require("prop-types");
prop_types = require_runtime.__toESM(prop_types);
let react_jsx_runtime = require("react/jsx-runtime");
let _carbon_icons_react = require("@carbon/icons-react");
//#region src/components/UIShell/SideNavSwitcher.tsx
/**
* Copyright IBM Corp. 2016, 2023
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
const SideNavSwitcher = (0, react.forwardRef)((props, ref) => {
	const id = require_useId.useId("side-nav-switcher");
	const prefix = require_usePrefix.usePrefix();
	const { className: customClassName, labelText, onChange, options } = props;
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
		className: (0, classnames.default)(`${prefix}--side-nav__switcher`, customClassName),
		children: [
			/* @__PURE__ */ (0, react_jsx_runtime.jsx)("label", {
				htmlFor: id,
				className: `${prefix}--assistive-text`,
				children: labelText
			}),
			/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("select", {
				id,
				className: `${prefix}--side-nav__select`,
				defaultValue: "",
				onBlur: onChange,
				onChange,
				ref,
				children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("option", {
					className: `${prefix}--side-nav__option`,
					disabled: true,
					hidden: true,
					value: "",
					children: labelText
				}), options.map((option) => /* @__PURE__ */ (0, react_jsx_runtime.jsx)("option", {
					className: `${prefix}--side-nav__option`,
					value: option,
					children: option
				}, option))]
			}),
			/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
				className: `${prefix}--side-nav__switcher-chevron`,
				children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(_carbon_icons_react.ChevronDown, { size: 20 })
			})
		]
	});
});
SideNavSwitcher.displayName = "SideNavSwitcher";
SideNavSwitcher.propTypes = {
	className: prop_types.default.string,
	labelText: prop_types.default.string.isRequired,
	onChange: prop_types.default.func,
	options: prop_types.default.arrayOf(prop_types.default.string).isRequired
};
//#endregion
exports.default = SideNavSwitcher;
