// @carbon/react/lib/components/RadioButtonGroup/RadioButtonGroup.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

const require_runtime = require("../../_virtual/_rolldown/runtime.js");
const require_usePrefix = require("../../internal/usePrefix.js");
const require_useId = require("../../internal/useId.js");
const require_deprecate = require("../../prop-types/deprecate.js");
const require_utils = require("../../internal/utils.js");
const require_hasHelperText = require("../../internal/hasHelperText.js");
const require_index = require("../AILabel/index.js");
const require_mergeRefs = require("../../tools/mergeRefs.js");
const require_index$1 = require("../RadioButton/index.js");
const require_createTextComponent = require("../Text/createTextComponent.js");
let classnames = require("classnames");
classnames = require_runtime.__toESM(classnames);
let react = require("react");
react = require_runtime.__toESM(react);
let prop_types = require("prop-types");
prop_types = require_runtime.__toESM(prop_types);
let react_jsx_runtime = require("react/jsx-runtime");
let _carbon_icons_react = require("@carbon/icons-react");
(0, react.createContext)(null);
const RadioButtonGroup = react.default.forwardRef((props, ref) => {
	const { children, className, decorator, defaultSelected, disabled, helperText, invalid = false, invalidText, labelPosition = "right", legendText, name, onChange = () => {}, orientation = "horizontal", readOnly, valueSelected, warn = false, warnText, slug, required, ...rest } = props;
	const prefix = require_usePrefix.usePrefix();
	const [selected, setSelected] = (0, react.useState)(valueSelected ?? defaultSelected);
	const prevValueSelected = (0, react.useRef)(valueSelected);
	const radioButtonGroupInstanceId = require_useId.useId();
	(0, react.useEffect)(() => {
		if (valueSelected !== prevValueSelected.current) {
			setSelected(valueSelected);
			prevValueSelected.current = valueSelected;
		}
	}, [valueSelected]);
	function getRadioButtons() {
		return react.Children.map(children, (radioButton) => {
			if (!require_utils.isComponentElement(radioButton, require_index$1.default)) return radioButton;
			const newProps = {
				name,
				key: radioButton.props.value,
				value: radioButton.props.value,
				onChange: handleOnChange,
				checked: radioButton.props.value === selected,
				required
			};
			if (!selected && radioButton.props.checked) newProps.checked = true;
			return (0, react.cloneElement)(radioButton, newProps);
		});
	}
	function handleOnChange(newSelection, value, evt) {
		if (!readOnly) {
			if (newSelection !== selected) {
				setSelected(newSelection);
				onChange(newSelection, name, evt);
			}
		}
	}
	const showWarning = !readOnly && !disabled && !invalid && warn;
	const showHelper = !invalid && !disabled && !warn;
	const wrapperClasses = (0, classnames.default)(`${prefix}--form-item`, className);
	const fieldsetClasses = (0, classnames.default)(`${prefix}--radio-button-group`, {
		[`${prefix}--radio-button-group--${orientation}`]: orientation === "vertical",
		[`${prefix}--radio-button-group--label-${labelPosition}`]: labelPosition,
		[`${prefix}--radio-button-group--readonly`]: readOnly,
		[`${prefix}--radio-button-group--invalid`]: !readOnly && !disabled && invalid,
		[`${prefix}--radio-button-group--warning`]: showWarning,
		[`${prefix}--radio-button-group--slug`]: slug,
		[`${prefix}--radio-button-group--decorator`]: decorator
	});
	const helperClasses = (0, classnames.default)(`${prefix}--form__helper-text`, { [`${prefix}--form__helper-text--disabled`]: disabled });
	const hasHelper = require_hasHelperText.hasHelperText(helperText);
	const helperId = !hasHelper ? void 0 : `radio-button-group-helper-text-${radioButtonGroupInstanceId}`;
	const helper = hasHelper && /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
		id: helperId,
		className: helperClasses,
		children: helperText
	});
	const divRef = (0, react.useRef)(null);
	const candidate = slug ?? decorator;
	const normalizedDecorator = require_utils.isComponentElement(candidate, require_index.AILabel) ? (0, react.cloneElement)(candidate, {
		size: "mini",
		kind: "default"
	}) : candidate;
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
		className: wrapperClasses,
		ref: require_mergeRefs.mergeRefs(divRef, ref),
		children: [
			/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("fieldset", {
				className: fieldsetClasses,
				disabled,
				"data-invalid": invalid ? true : void 0,
				"aria-describedby": showHelper && hasHelper ? helperId : void 0,
				...rest,
				children: [legendText && /* @__PURE__ */ (0, react_jsx_runtime.jsxs)(require_createTextComponent.Legend, {
					className: `${prefix}--label`,
					children: [legendText, slug ? normalizedDecorator : decorator ? /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
						className: `${prefix}--radio-button-group-inner--decorator`,
						children: normalizedDecorator
					}) : ""]
				}), getRadioButtons()]
			}),
			/* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
				className: `${prefix}--radio-button__validation-msg`,
				children: [!readOnly && !disabled && invalid && /* @__PURE__ */ (0, react_jsx_runtime.jsxs)(react_jsx_runtime.Fragment, { children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)(_carbon_icons_react.WarningFilled, { className: `${prefix}--radio-button__invalid-icon` }), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
					className: `${prefix}--form-requirement`,
					children: invalidText
				})] }), showWarning && /* @__PURE__ */ (0, react_jsx_runtime.jsxs)(react_jsx_runtime.Fragment, { children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)(_carbon_icons_react.WarningAltFilled, { className: `${prefix}--radio-button__invalid-icon ${prefix}--radio-button__invalid-icon--warning` }), /* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
					className: `${prefix}--form-requirement`,
					children: warnText
				})] })]
			}),
			showHelper && helper
		]
	});
});
RadioButtonGroup.propTypes = {
	children: prop_types.default.node,
	className: prop_types.default.string,
	decorator: prop_types.default.node,
	defaultSelected: prop_types.default.oneOfType([prop_types.default.string, prop_types.default.number]),
	disabled: prop_types.default.bool,
	helperText: prop_types.default.node,
	invalid: prop_types.default.bool,
	invalidText: prop_types.default.node,
	labelPosition: prop_types.default.oneOf(["left", "right"]),
	legendText: prop_types.default.node,
	name: prop_types.default.string.isRequired,
	onChange: prop_types.default.func,
	orientation: prop_types.default.oneOf(["horizontal", "vertical"]),
	readOnly: prop_types.default.bool,
	required: prop_types.default.bool,
	slug: require_deprecate.deprecate(prop_types.default.node, "The `slug` prop has been deprecated and will be removed in the next major version. Use the decorator prop instead."),
	valueSelected: prop_types.default.oneOfType([prop_types.default.string, prop_types.default.number]),
	warn: prop_types.default.bool,
	warnText: prop_types.default.node
};
RadioButtonGroup.displayName = "RadioButtonGroup";
//#endregion
exports.default = RadioButtonGroup;
