// @carbon/react/lib/components/ContainedList/ContainedList.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

const require_runtime = require("../../_virtual/_rolldown/runtime.js");
const require_usePrefix = require("../../internal/usePrefix.js");
const require_useId = require("../../internal/useId.js");
const require_utils = require("../../internal/utils.js");
const require_index = require("../Layout/index.js");
const require_index$1 = require("./ContainedListItem/index.js");
const require_Search = require("../Search/Search.js");
const require_index$2 = require("../ExpandableSearch/index.js");
let classnames = require("classnames");
classnames = require_runtime.__toESM(classnames);
let react = require("react");
react = require_runtime.__toESM(react);
let prop_types = require("prop-types");
prop_types = require_runtime.__toESM(prop_types);
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/components/ContainedList/ContainedList.tsx
/**
* Copyright IBM Corp. 2022, 2026
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
const variants = ["on-page", "disclosed"];
const isSearchComponent = (node) => require_utils.isComponentElement(node, require_Search.default) || require_utils.isComponentElement(node, require_index$2.default);
const filterChildren = (children) => {
	if (Array.isArray(children)) return children.filter((child) => !isSearchComponent(child));
	if (children && !isSearchComponent(children)) return children;
	return null;
};
const ContainedList = ({ action, children, className, isInset, kind = variants[0], label, size, ...rest }) => {
	const labelId = `${require_useId.useId("contained-list")}-header`;
	const prefix = require_usePrefix.usePrefix();
	const classes = (0, classnames.default)(`${prefix}--contained-list`, {
		[`${prefix}--contained-list--inset-rulers`]: isInset,
		[`${prefix}--contained-list--${size}`]: size,
		[`${prefix}--layout--size-${size}`]: size
	}, `${prefix}--contained-list--${kind}`, className);
	const filteredChildren = filterChildren(children);
	const isActionSearch = isSearchComponent(action);
	return /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
		className: classes,
		...rest,
		children: [label && /* @__PURE__ */ (0, react_jsx_runtime.jsxs)("div", {
			className: `${prefix}--contained-list__header`,
			children: [/* @__PURE__ */ (0, react_jsx_runtime.jsx)("div", {
				id: labelId,
				className: `${prefix}--contained-list__label`,
				children: label
			}), /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_index.LayoutConstraint, {
				size: {
					min: "sm",
					max: "xl"
				},
				className: `${prefix}--contained-list__action`,
				children: action
			})]
		}), children && /* @__PURE__ */ (0, react_jsx_runtime.jsx)(
			"ul",
			/**
			* Webkit removes implicit "list" semantics when "list-style-type: none" is set.
			* Explicitly setting the "list" role ensures assistive technology in webkit
			* browsers correctly announce the semantics.
			*
			* Ref https://bugs.webkit.org/show_bug.cgi?id=170179#c1
			*/
			{
				role: "list",
				"aria-labelledby": label ? labelId : void 0,
				children: isActionSearch ? filteredChildren : children
			}
		)]
	});
};
ContainedList.propTypes = {
	action: prop_types.default.node,
	children: prop_types.default.node,
	className: prop_types.default.string,
	isInset: prop_types.default.bool,
	kind: prop_types.default.oneOf(variants),
	label: prop_types.default.oneOfType([prop_types.default.string, prop_types.default.node]),
	size: prop_types.default.oneOf([
		"sm",
		"md",
		"lg",
		"xl"
	])
};
Object.assign(ContainedList, { ContainedListItem: require_index$1.default });
//#endregion
exports.default = ContainedList;
