// @carbon/react/lib/components/DataTable/TableSelectRow.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

const require_runtime = require("../../_virtual/_rolldown/runtime.js");
const require_usePrefix = require("../../internal/usePrefix.js");
const require_useId = require("../../internal/useId.js");
const require_deprecate = require("../../prop-types/deprecate.js");
const require_index = require("../InlineCheckbox/index.js");
const require_index$1 = require("../RadioButton/index.js");
let classnames = require("classnames");
classnames = require_runtime.__toESM(classnames);
let react = require("react");
react = require_runtime.__toESM(react);
let prop_types = require("prop-types");
prop_types = require_runtime.__toESM(prop_types);
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/components/DataTable/TableSelectRow.tsx
/**
* Copyright IBM Corp. 2016, 2025
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
const TableSelectRow = ({ ariaLabel: deprecatedAriaLabel, ["aria-label"]: ariaLabel, checked, id, name, onSelect, onChange, disabled, radio, className }) => {
	const prefix = require_usePrefix.usePrefix();
	const uniqueNameId = require_useId.useId();
	const handleRadioChange = onChange ? (value, name, event) => {
		onChange(!!value, name || "", event);
	} : void 0;
	const handleCheckboxChange = onChange ? (checked, name, event) => {
		onChange(checked, name, event);
	} : void 0;
	const selectionInputProps = {
		id,
		name: name ? name : uniqueNameId,
		onClick: onSelect,
		checked,
		disabled
	};
	const labelValue = ariaLabel || deprecatedAriaLabel || "";
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("td", {
		className: (0, classnames.default)(`${prefix}--table-column-checkbox`, {
			...className && { [className]: true },
			[`${prefix}--table-column-radio`]: radio
		}),
		"aria-live": "off",
		children: radio ? /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_index$1.default, {
			...selectionInputProps,
			labelText: labelValue,
			onChange: handleRadioChange,
			hideLabel: true
		}) : /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_index.default, {
			...selectionInputProps,
			"aria-label": labelValue,
			onChange: handleCheckboxChange
		})
	});
};
TableSelectRow.propTypes = {
	["aria-label"]: prop_types.default.string,
	ariaLabel: require_deprecate.deprecate(prop_types.default.string, "This prop syntax has been deprecated. Please use the new `aria-label`."),
	checked: prop_types.default.bool,
	className: prop_types.default.string,
	disabled: prop_types.default.bool,
	id: prop_types.default.string.isRequired,
	name: prop_types.default.string.isRequired,
	onChange: prop_types.default.func,
	onSelect: prop_types.default.func.isRequired,
	radio: prop_types.default.bool
};
//#endregion
exports.default = TableSelectRow;
