// @carbon/react/lib/components/DataTable/TableToolbarMenu.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

const require_runtime = require("../../_virtual/_rolldown/runtime.js");
const require_usePrefix = require("../../internal/usePrefix.js");
const require_TableToolbar = require("./TableToolbar.js");
const require_index = require("../OverflowMenu/index.js");
let classnames = require("classnames");
classnames = require_runtime.__toESM(classnames);
let react = require("react");
react = require_runtime.__toESM(react);
let prop_types = require("prop-types");
prop_types = require_runtime.__toESM(prop_types);
let react_jsx_runtime = require("react/jsx-runtime");
let _carbon_icons_react = require("@carbon/icons-react");
//#region src/components/DataTable/TableToolbarMenu.tsx
/**
* Copyright IBM Corp. 2016, 2026
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
const defaultIconDescription = "Settings";
const TableToolbarMenu = ({ className, renderIcon = _carbon_icons_react.Settings, iconDescription = defaultIconDescription, children, menuOptionsClass, size: sizeProp, ...rest }) => {
	const toolbarContext = require_TableToolbar.useTableToolbar();
	const size = sizeProp ?? toolbarContext.size;
	const prefix = require_usePrefix.usePrefix();
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_index.default, {
		renderIcon,
		className: (0, classnames.default)(className, `${prefix}--toolbar-action ${prefix}--overflow-menu`),
		iconDescription,
		menuOptionsClass: (0, classnames.default)(menuOptionsClass, `${prefix}--toolbar-action__menu`),
		size,
		flipped: true,
		...rest,
		children
	});
};
TableToolbarMenu.propTypes = {
	children: prop_types.default.node.isRequired,
	className: prop_types.default.string,
	iconDescription: prop_types.default.string,
	menuOptionsClass: prop_types.default.string,
	renderIcon: prop_types.default.oneOfType([prop_types.default.func, prop_types.default.object]),
	size: prop_types.default.oneOf([
		"xs",
		"sm",
		"md",
		"lg"
	])
};
//#endregion
exports.default = TableToolbarMenu;
