// @carbon/react/lib/components/DataTable/TableSelectAll.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

const require_runtime = require("../../_virtual/_rolldown/runtime.js");
const require_usePrefix = require("../../internal/usePrefix.js");
const require_deprecate = require("../../prop-types/deprecate.js");
const require_index = require("../InlineCheckbox/index.js");
let classnames = require("classnames");
classnames = require_runtime.__toESM(classnames);
let react = require("react");
react = require_runtime.__toESM(react);
let prop_types = require("prop-types");
prop_types = require_runtime.__toESM(prop_types);
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/components/DataTable/TableSelectAll.tsx
/**
* Copyright IBM Corp. 2016, 2025
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
const TableSelectAll = ({ ariaLabel: deprecatedAriaLabel = "Select all rows in the table", ["aria-label"]: ariaLabel, checked, id, indeterminate, name, onSelect, disabled, className }) => {
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)("th", {
		"aria-live": "off",
		scope: "col",
		className: (0, classnames.default)(`${require_usePrefix.usePrefix()}--table-column-checkbox`, className),
		children: /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_index.default, {
			"aria-label": ariaLabel || deprecatedAriaLabel,
			checked,
			id,
			indeterminate,
			name,
			onClick: onSelect,
			disabled
		})
	});
};
TableSelectAll.propTypes = {
	["aria-label"]: prop_types.default.string,
	ariaLabel: require_deprecate.deprecate(prop_types.default.string, "This prop syntax has been deprecated. Please use the new `aria-label`."),
	checked: prop_types.default.bool,
	className: prop_types.default.string,
	disabled: prop_types.default.bool,
	id: prop_types.default.string.isRequired,
	indeterminate: prop_types.default.bool,
	name: prop_types.default.string.isRequired,
	onSelect: prop_types.default.func.isRequired
};
//#endregion
exports.default = TableSelectAll;
