// @carbon/react/lib/components/Pagination/experimental/PageSelector.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

const require_runtime = require("../../../_virtual/_rolldown/runtime.js");
const require_usePrefix = require("../../../internal/usePrefix.js");
const require_useId = require("../../../internal/useId.js");
const require_index = require("../../Select/index.js");
const require_index$1 = require("../../SelectItem/index.js");
let classnames = require("classnames");
classnames = require_runtime.__toESM(classnames);
let react = require("react");
react = require_runtime.__toESM(react);
let prop_types = require("prop-types");
prop_types = require_runtime.__toESM(prop_types);
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/components/Pagination/experimental/PageSelector.js
/**
* Copyright IBM Corp. 2016, 2023
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
function PageSelector({ className = null, currentPage, id = 1, labelText = "Current page number", totalPages, ...other }) {
	const namespace = `${require_usePrefix.usePrefix()}--unstable-pagination__page-selector`;
	const instanceId = `${namespace}__select-${require_useId.useId()}`;
	const renderPages = (total) => {
		const pages = [];
		for (let counter = 1; counter <= total; counter += 1) pages.push(/* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_index$1.default, {
			value: counter,
			text: String(counter)
		}, counter));
		return pages;
	};
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(require_index.default, {
		className: (0, classnames.default)(namespace, className),
		hideLabel: true,
		id: instanceId || id,
		inline: true,
		labelText,
		value: currentPage,
		...other,
		children: renderPages(totalPages)
	});
}
PageSelector.propTypes = {
	className: prop_types.default.string,
	currentPage: prop_types.default.number.isRequired,
	id: prop_types.default.oneOfType([prop_types.default.string, prop_types.default.number]),
	labelText: prop_types.default.string,
	totalPages: prop_types.default.number.isRequired
};
//#endregion
exports.default = PageSelector;
