// @carbon/react/lib/components/Layout/index.js
/**
 * Copyright IBM Corp. 2016, 2026
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

const require_runtime = require("../../_virtual/_rolldown/runtime.js");
const require_usePrefix = require("../../internal/usePrefix.js");
let classnames = require("classnames");
classnames = require_runtime.__toESM(classnames);
let react = require("react");
react = require_runtime.__toESM(react);
let prop_types = require("prop-types");
prop_types = require_runtime.__toESM(prop_types);
let react_jsx_runtime = require("react/jsx-runtime");
//#region src/components/Layout/index.tsx
/**
* Copyright IBM Corp. 2023
*
* This source code is licensed under the Apache-2.0 license found in the
* LICENSE file in the root directory of this source tree.
*/
const sizes = [
	"xs",
	"sm",
	"md",
	"lg",
	"xl",
	"2xl"
];
const densities = ["condensed", "normal"];
const Layout = react.default.forwardRef(({ as: BaseComponent = "div", children, className, density, size, ...rest }, forwardRef) => {
	const prefix = require_usePrefix.usePrefix();
	const classes = (0, classnames.default)(className, `${prefix}--layout`, {
		[`${prefix}--layout--size-${size}`]: size && sizes.includes(size),
		[`${prefix}--layout--density-${density}`]: density && densities.includes(density)
	});
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(BaseComponent, {
		...rest,
		ref: forwardRef,
		className: classes,
		children
	});
});
Layout.propTypes = {
	as: prop_types.default.oneOfType([
		prop_types.default.func,
		prop_types.default.string,
		prop_types.default.elementType
	]),
	children: prop_types.default.node,
	className: prop_types.default.string,
	density: prop_types.default.oneOf(densities),
	size: prop_types.default.oneOf(sizes)
};
const LayoutConstraint = react.default.forwardRef(({ as: BaseComponent = "div", children, className, density, size, ...rest }, forwardRef) => {
	const prefix = require_usePrefix.usePrefix();
	const classes = (0, classnames.default)(className, Object.entries({
		size,
		density
	}).map(([group, constraints]) => ({
		[`${prefix}--layout-constraint--${group}__default-${constraints?.default}`]: constraints?.default,
		[`${prefix}--layout-constraint--${group}__min-${constraints?.min}`]: constraints?.min,
		[`${prefix}--layout-constraint--${group}__max-${constraints?.max}`]: constraints?.max
	})));
	return /* @__PURE__ */ (0, react_jsx_runtime.jsx)(BaseComponent, {
		...rest,
		ref: forwardRef,
		className: classes,
		children
	});
});
LayoutConstraint.propTypes = {
	as: prop_types.default.oneOfType([
		prop_types.default.func,
		prop_types.default.string,
		prop_types.default.elementType
	]),
	children: prop_types.default.node,
	className: prop_types.default.string,
	density: prop_types.default.shape({
		min: prop_types.default.oneOf(densities),
		default: prop_types.default.oneOf(densities),
		max: prop_types.default.oneOf(densities)
	}),
	size: prop_types.default.shape({
		min: prop_types.default.oneOf(sizes),
		default: prop_types.default.oneOf(sizes),
		max: prop_types.default.oneOf(sizes)
	})
};
//#endregion
exports.Layout = Layout;
exports.LayoutConstraint = LayoutConstraint;
